// xrm-mock/build/index.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
define("xrm-mock/controls/addcontrolnotificationoptions/addcontrolnotificationoptions.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AddControlNotificationOptionsMock = /** @class */ (function () {
        function AddControlNotificationOptionsMock(components) {
            this.actions = components.actions;
            this.messages = components.messages;
            this.notificationLevel = components.notificationLevel;
            this.uniqueId = components.uniqueId;
        }
        return AddControlNotificationOptionsMock;
    }());
    exports.AddControlNotificationOptionsMock = AddControlNotificationOptionsMock;
});
define("xrm-mock/navigation/alertstrings/alertstrings.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AlertStringsMock = /** @class */ (function () {
        function AlertStringsMock(text, confirmButtonLabel) {
            this.text = text;
            this.confirmButtonLabel = confirmButtonLabel;
        }
        return AlertStringsMock;
    }());
    exports.AlertStringsMock = AlertStringsMock;
});
define("xrm-mock/appproperties/appproperties.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AppPropertiesMock = /** @class */ (function () {
        function AppPropertiesMock(components) {
            this.appId = components.appId;
            this.displayName = components.displayName;
            this.uniqueName = components.uniqueName;
            this.url = components.url;
            this.webResourceId = components.webResourceId;
            this.webResourceName = components.webResourceName;
            this.welcomePageId = components.welcomePageId;
            this.welcomePageName = components.welcomePageName;
        }
        return AppPropertiesMock;
    }());
    exports.AppPropertiesMock = AppPropertiesMock;
});
define("xrm-mock/metadata/attributemetadata/attributemetadata.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AttributeMetadataMock = /** @class */ (function () {
        function AttributeMetadataMock(components) {
            this.defaultFormValue = components.defaultFormValue;
            this.options = components.options;
            this.logicalName = components.logicalName;
            this.displayName = components.displayName;
            this.attributeType = components.attributeType;
            this.entityLogicalName = components.entityLogicalName;
            this.optionSet = components.optionSet;
        }
        return AttributeMetadataMock;
    }());
    exports.AttributeMetadataMock = AttributeMetadataMock;
});
define("xrm-mock-generator/helpers/array.helper", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    function findIndex(handlers, handler) {
        return handlers.findIndex(function (item) { return item.name === handler.name; });
    }
    exports.findIndex = findIndex;
});
define("xrm-mock/collection/itemcollection/itemcollection.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ItemCollectionMock = /** @class */ (function () {
        function ItemCollectionMock(itemCollection) {
            this.itemCollection = itemCollection || [];
        }
        ItemCollectionMock.prototype.forEach = function (delegate) {
            this.itemCollection.map(delegate);
        };
        ItemCollectionMock.prototype.get = function (param) {
            if (param === undefined || param === null) {
                return this.itemCollection;
            }
            else if (typeof param === "string") {
                var attribute = void 0;
                for (var _i = 0, _a = this.itemCollection; _i < _a.length; _i++) {
                    var item = _a[_i];
                    if (item.getName !== undefined && item.getName() === param) {
                        attribute = item;
                        break;
                    }
                }
                return attribute || null;
            }
            else if (typeof param === "number") {
                return this.itemCollection[param] || null;
            }
            else {
                return this.itemCollection.filter(param);
            }
        };
        ItemCollectionMock.prototype.getLength = function () {
            return this.itemCollection.length;
        };
        ItemCollectionMock.prototype.push = function (item) {
            this.itemCollection.push(item);
        };
        return ItemCollectionMock;
    }());
    exports.ItemCollectionMock = ItemCollectionMock;
});
define("xrm-mock/controls/control/control.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ControlMock = /** @class */ (function () {
        function ControlMock(components) {
            this.controlType = components.controlType || "standard";
            this.label = components.label;
            this.visible = components.visible;
            this.name = components.name;
            this.parent = components.parent;
        }
        ControlMock.prototype.getControlType = function () {
            return this.controlType;
        };
        ControlMock.prototype.getName = function () {
            return this.name;
        };
        ControlMock.prototype.getParent = function () {
            return this.parent;
        };
        ControlMock.prototype.getLabel = function () {
            return this.label;
        };
        ControlMock.prototype.setLabel = function (label) {
            this.label = label;
        };
        ControlMock.prototype.getVisible = function () {
            return this.visible;
        };
        return ControlMock;
    }());
    exports.ControlMock = ControlMock;
});
define("xrm-mock/attributes/attribute/attribute.mock", ["require", "exports", "xrm-mock-generator/helpers/array.helper", "xrm-mock/collection/itemcollection/itemcollection.mock"], function (require, exports, array_helper_1, itemcollection_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AttributeMock = /** @class */ (function () {
        function AttributeMock(components) {
            this.attributeType = components.attributeType || "string";
            this.format = components.format;
            this.name = components.name;
            this.value = components.value;
            this.isDirty = components.isDirty === true ? true : false;
            this.requiredLevel = components.requiredLevel || "none";
            this.submitMode = components.submitMode || "dirty";
            this.controls = components.controls || new itemcollection_mock_1.ItemCollectionMock();
            this.eventHandlers = components.eventHandlers || [];
        }
        AttributeMock.prototype.addOnChange = function (handler) {
            this.eventHandlers.push(handler);
        };
        AttributeMock.prototype.fireOnChange = function () {
            if (this.eventHandlers.length) {
                for (var _i = 0, _a = this.eventHandlers; _i < _a.length; _i++) {
                    var handler = _a[_i];
                    handler.call(this);
                }
            }
        };
        AttributeMock.prototype.getAttributeType = function () {
            return this.attributeType;
        };
        AttributeMock.prototype.getFormat = function () {
            return this.format;
        };
        AttributeMock.prototype.getIsDirty = function () {
            return this.isDirty;
        };
        AttributeMock.prototype.getName = function () {
            return this.name;
        };
        AttributeMock.prototype.getParent = function () {
            throw new Error("getParent not implemented");
        };
        AttributeMock.prototype.getRequiredLevel = function () {
            return this.requiredLevel;
        };
        AttributeMock.prototype.getSubmitMode = function () {
            return this.submitMode;
        };
        AttributeMock.prototype.getUserPrivilege = function () {
            throw new Error("getUserPrivilege not implemented");
        };
        AttributeMock.prototype.removeOnChange = function (handler) {
            var index = array_helper_1.findIndex(this.eventHandlers, handler);
            this.eventHandlers.splice(index, 1);
        };
        AttributeMock.prototype.setRequiredLevel = function (requirementLevel) {
            this.requiredLevel = requirementLevel;
        };
        AttributeMock.prototype.setSubmitMode = function (submitMode) {
            this.submitMode = submitMode;
        };
        AttributeMock.prototype.getValue = function () {
            return this.value;
        };
        AttributeMock.prototype.setValue = function (value) {
            this.value = value;
            this.isDirty = true;
        };
        return AttributeMock;
    }());
    exports.AttributeMock = AttributeMock;
});
define("xrm-mock/controls/autocompletecommand/autocompletecommand.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AutoCompleteCommandMock = /** @class */ (function () {
        function AutoCompleteCommandMock(id, label, icon, action) {
            this.id = id;
            this.label = label;
            this.icon = icon;
            this.action = action;
        }
        AutoCompleteCommandMock.prototype.action = function () {
            throw new Error("Action not implemented.");
        };
        return AutoCompleteCommandMock;
    }());
    exports.AutoCompleteCommandMock = AutoCompleteCommandMock;
});
define("xrm-mock/controls/autocompleteresult/autocompleteresult.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AutoCompleteResultMock = /** @class */ (function () {
        function AutoCompleteResultMock(id, fields, icon) {
            this.id = id;
            this.fields = fields;
            this.icon = icon;
        }
        return AutoCompleteResultMock;
    }());
    exports.AutoCompleteResultMock = AutoCompleteResultMock;
});
define("xrm-mock/controls/autocompleteresultset/autocompleteresultset.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AutoCompleteResultSetMock = /** @class */ (function () {
        function AutoCompleteResultSetMock(results, commands) {
            this.results = results;
            this.commands = commands;
        }
        return AutoCompleteResultSetMock;
    }());
    exports.AutoCompleteResultSetMock = AutoCompleteResultSetMock;
});
define("xrm-mock/controls/uifocusable/uifocusable.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var UiFocusableMock = /** @class */ (function () {
        function UiFocusableMock(hasFocus) {
            if (hasFocus === void 0) { hasFocus = false; }
            this.hasFocus = hasFocus;
        }
        UiFocusableMock.prototype.setFocus = function () {
            this.hasFocus = true;
        };
        return UiFocusableMock;
    }());
    exports.UiFocusableMock = UiFocusableMock;
});
define("xrm-mock/controls/uicangetvisibleelement/uicangetvisibleelement.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var UiCanGetVisibleElementMock = /** @class */ (function () {
        function UiCanGetVisibleElementMock(isVisible) {
            this.isVisible = isVisible;
        }
        UiCanGetVisibleElementMock.prototype.getVisible = function () {
            return this.isVisible;
        };
        return UiCanGetVisibleElementMock;
    }());
    exports.UiCanGetVisibleElementMock = UiCanGetVisibleElementMock;
});
define("xrm-mock/controls/uilabelelement/uilabelelement.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var UiLabelElementMock = /** @class */ (function () {
        function UiLabelElementMock(label) {
            this.label = label;
        }
        UiLabelElementMock.prototype.getLabel = function () {
            return this.label;
        };
        UiLabelElementMock.prototype.setLabel = function (label) {
            this.label = label;
        };
        return UiLabelElementMock;
    }());
    exports.UiLabelElementMock = UiLabelElementMock;
});
define("xrm-mock/controls/uistandardelement/uistandardelement.mock", ["require", "exports", "xrm-mock/controls/uicangetvisibleelement/uicangetvisibleelement.mock", "xrm-mock/controls/uilabelelement/uilabelelement.mock"], function (require, exports, uicangetvisibleelement_mock_1, uilabelelement_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var UiStandardElementMock = /** @class */ (function () {
        function UiStandardElementMock(uiLabelElement, uiCanGetVisibleElement) {
            this.uiLabelElement = uiLabelElement;
            this.uiCanGetVisibleElement = uiCanGetVisibleElement;
        }
        UiStandardElementMock.create = function (label, visible) {
            if (visible === void 0) { visible = true; }
            return new UiStandardElementMock(new uilabelelement_mock_1.UiLabelElementMock(label), new uicangetvisibleelement_mock_1.UiCanGetVisibleElementMock(visible));
        };
        UiStandardElementMock.prototype.setVisible = function (visible) {
            this.uiCanGetVisibleElement.getVisible = function () { return visible; };
        };
        UiStandardElementMock.prototype.getVisible = function () {
            return this.uiCanGetVisibleElement.getVisible();
        };
        UiStandardElementMock.prototype.getLabel = function () {
            return this.uiLabelElement.getLabel();
        };
        UiStandardElementMock.prototype.setLabel = function (label) {
            this.uiLabelElement.setLabel(label);
        };
        return UiStandardElementMock;
    }());
    exports.UiStandardElementMock = UiStandardElementMock;
});
define("xrm-mock/controls/standardcontrol/standardcontrol.mock", ["require", "exports", "xrm-mock/controls/control/control.mock", "xrm-mock/controls/uifocusable/uifocusable.mock", "xrm-mock/controls/uistandardelement/uistandardelement.mock"], function (require, exports, control_mock_1, uifocusable_mock_1, uistandardelement_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StandardControlMock = /** @class */ (function (_super) {
        __extends(StandardControlMock, _super);
        function StandardControlMock(components) {
            var _this = _super.call(this, components) || this;
            _this.disabled = components.disabled || false;
            _this.uiStandardElement = uistandardelement_mock_1.UiStandardElementMock.create(components.label || _this.name, components.visible);
            _this.uiFocusable = new uifocusable_mock_1.UiFocusableMock(components.hasFocus);
            _this.attribute = components.attribute;
            if (_this.attribute && _this.attribute.controls) {
                _this.attribute.controls.push(_this);
            }
            return _this;
        }
        StandardControlMock.prototype.addNotification = function (notification) {
            throw new Error("Method not implemented.");
        };
        StandardControlMock.prototype.clearNotification = function (uniqueId) {
            throw new Error(("clear notification not implemented"));
        };
        StandardControlMock.prototype.getDisabled = function () {
            return this.disabled;
        };
        StandardControlMock.prototype.setDisabled = function (disabled) {
            this.disabled = disabled;
        };
        StandardControlMock.prototype.setNotification = function (message, uniqueId) {
            throw new Error(("set notification not implemented"));
        };
        StandardControlMock.prototype.getAttribute = function () {
            if (["subgrid", "iframe", "webresource"].indexOf(this.getControlType()) === -1) {
                return this.attribute;
            }
            else {
                throw new Error("control is not bound to an attribute because it is of type " + this.getControlType());
            }
        };
        StandardControlMock.prototype.getLabel = function () {
            return this.uiStandardElement.getLabel();
        };
        StandardControlMock.prototype.setLabel = function (label) {
            this.uiStandardElement.setLabel(label);
        };
        StandardControlMock.prototype.getVisible = function () {
            return this.uiStandardElement.getVisible();
        };
        StandardControlMock.prototype.setVisible = function (visible) {
            this.uiStandardElement.setVisible(visible);
        };
        StandardControlMock.prototype.setFocus = function () {
            this.uiFocusable.setFocus();
        };
        return StandardControlMock;
    }(control_mock_1.ControlMock));
    exports.StandardControlMock = StandardControlMock;
});
define("xrm-mock/controls/uikeypressable/uikeypressable.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var UiKeyPressableMock = /** @class */ (function () {
        function UiKeyPressableMock(keyPressHandlers) {
            this.keyPressHandlers = keyPressHandlers;
            this.keyPressHandlers = keyPressHandlers || [];
        }
        UiKeyPressableMock.prototype.addOnKeyPress = function (handler) {
            this.keyPressHandlers.push(handler);
        };
        UiKeyPressableMock.prototype.fireOnKeyPress = function (eventContext) {
            this.keyPressHandlers.forEach(function (k) {
                k(eventContext);
            });
        };
        UiKeyPressableMock.prototype.removeOnKeyPress = function (handler) {
            var index = this.keyPressHandlers.indexOf(handler);
            if (index > -1) {
                this.keyPressHandlers.splice(index);
            }
        };
        return UiKeyPressableMock;
    }());
    exports.UiKeyPressableMock = UiKeyPressableMock;
});
define("xrm-mock/controls/autolookupcontrol/autolookupcontrol.mock", ["require", "exports", "xrm-mock/controls/standardcontrol/standardcontrol.mock", "xrm-mock/controls/uikeypressable/uikeypressable.mock"], function (require, exports, standardcontrol_mock_1, uikeypressable_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AutoLookupControlMock = /** @class */ (function (_super) {
        __extends(AutoLookupControlMock, _super);
        function AutoLookupControlMock(components) {
            var _this = _super.call(this, components) || this;
            _this.uncommittedText = components.uncommittedText;
            _this.uiKeyPressable = new uikeypressable_mock_1.UiKeyPressableMock(components.keyPressHandlers);
            return _this;
        }
        AutoLookupControlMock.prototype.getValue = function () {
            return this.uncommittedText;
        };
        AutoLookupControlMock.prototype.hideAutoComplete = function () {
            throw new Error(("hide autocomplete not implemented"));
        };
        AutoLookupControlMock.prototype.showAutoComplete = function (resultSet) {
            throw new Error(("show autocomplete not implemented"));
        };
        AutoLookupControlMock.prototype.addOnKeyPress = function (handler) {
            this.uiKeyPressable.addOnKeyPress(handler);
        };
        AutoLookupControlMock.prototype.fireOnKeyPress = function () {
            this.uiKeyPressable.fireOnKeyPress();
        };
        AutoLookupControlMock.prototype.removeOnKeyPress = function (handler) {
            this.uiKeyPressable.removeOnKeyPress(handler);
        };
        return AutoLookupControlMock;
    }(standardcontrol_mock_1.StandardControlMock));
    exports.AutoLookupControlMock = AutoLookupControlMock;
});
define("xrm-mock/controls/booleancontrol/booleancontrol.mock", ["require", "exports", "xrm-mock/controls/standardcontrol/standardcontrol.mock"], function (require, exports, standardcontrol_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var BooleanControlMock = /** @class */ (function (_super) {
        __extends(BooleanControlMock, _super);
        function BooleanControlMock(components) {
            return _super.call(this, components) || this;
        }
        return BooleanControlMock;
    }(standardcontrol_mock_2.StandardControlMock));
    exports.BooleanControlMock = BooleanControlMock;
});
define("xrm-mock/attributes/enumattribute/enumattribute.mock", ["require", "exports", "xrm-mock/attributes/attribute/attribute.mock"], function (require, exports, attribute_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var EnumAttributeMock = /** @class */ (function (_super) {
        __extends(EnumAttributeMock, _super);
        function EnumAttributeMock(components) {
            var _this = _super.call(this, components) || this;
            _this.initialValue = "initialValue" in components
                ? components.initialValue
                : components.value;
            return _this;
        }
        EnumAttributeMock.prototype.getInitialValue = function () {
            return this.initialValue;
        };
        return EnumAttributeMock;
    }(attribute_mock_1.AttributeMock));
    exports.EnumAttributeMock = EnumAttributeMock;
});
define("xrm-mock/attributes/booleanattribute/booleanattribute.mock", ["require", "exports", "xrm-mock/attributes/enumattribute/enumattribute.mock"], function (require, exports, enumattribute_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var BooleanAttributeMock = /** @class */ (function (_super) {
        __extends(BooleanAttributeMock, _super);
        function BooleanAttributeMock(components) {
            return _super.call(this, BooleanAttributeMock.defaultComponents(components)) || this;
        }
        BooleanAttributeMock.create = function (name, value) {
            if (value === void 0) { value = false; }
            return new BooleanAttributeMock({ name: name, value: value });
        };
        BooleanAttributeMock.defaultComponents = function (components) {
            components.attributeType = "boolean";
            return components;
        };
        return BooleanAttributeMock;
    }(enumattribute_mock_1.EnumAttributeMock));
    exports.BooleanAttributeMock = BooleanAttributeMock;
});
define("xrm-mock/calendar/calendar.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var CalendarMock = /** @class */ (function () {
        function CalendarMock(components) {
            this.MinSupportedDateTime = components.MinSupportedDateTime;
            this.MaxSupportedDateTime = components.MaxSupportedDateTime;
            this.AlgorithmType = components.AlgorithmType;
            this.CalendarType = components.CalendarType;
            this.Eras = components.Eras;
            this.TwoDigitYearMax = components.TwoDigitYearMax;
            this.IsReadOnly = components.IsReadOnly;
        }
        return CalendarMock;
    }());
    exports.CalendarMock = CalendarMock;
});
define("xrm-mock/device/capturefileresponse/capturefileresponse.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var CaptureFileResponseMock = /** @class */ (function () {
        function CaptureFileResponseMock(fileContent, fileName, fileSize, mimeType) {
            this.fileContent = fileContent;
            this.fileName = fileName;
            this.fileSize = fileSize;
            this.mimeType = mimeType;
        }
        return CaptureFileResponseMock;
    }());
    exports.CaptureFileResponseMock = CaptureFileResponseMock;
});
define("xrm-mock/device/captureimageoptions/captureimageoptions.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var CaptureImageOptionsMock = /** @class */ (function () {
        function CaptureImageOptionsMock(components) {
            this.allowEdit = components.allowEdit;
            this.height = components.height;
            this.preferFrontCamera = components.preferFrontCamera;
            this.quality = components.quality;
            this.width = components.width;
        }
        return CaptureImageOptionsMock;
    }());
    exports.CaptureImageOptionsMock = CaptureImageOptionsMock;
});
define("xrm-mock/clientcontext/clientcontext.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ClientContextMock = /** @class */ (function () {
        function ClientContextMock(client, clientState) {
            this.client = client;
            this.clientState = clientState;
        }
        ClientContextMock.prototype.getClient = function () {
            return this.client;
        };
        ClientContextMock.prototype.getClientState = function () {
            return this.clientState;
        };
        ClientContextMock.prototype.getFormFactor = function () {
            throw new Error("Method not implemented.");
        };
        ClientContextMock.prototype.isOffline = function () {
            throw new Error("Method not implemented.");
        };
        return ClientContextMock;
    }());
    exports.ClientContextMock = ClientContextMock;
});
define("xrm-mock/navigation/confirmresult/confirmresult.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ConfirmResultMock = /** @class */ (function () {
        function ConfirmResultMock(confirmed) {
            this.confirmed = confirmed;
        }
        return ConfirmResultMock;
    }());
    exports.ConfirmResultMock = ConfirmResultMock;
});
define("xrm-mock/navigation/confirmstrings/confirmstrings.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ConfirmStringsMock = /** @class */ (function () {
        function ConfirmStringsMock(components) {
            this.cancelButtonLabel = components.cancelButtonLabel;
            this.confirmButtonLabel = components.confirmButtonLabel;
            this.subtitle = components.subtitle;
            this.title = components.title;
            this.text = components.text;
        }
        return ConfirmStringsMock;
    }());
    exports.ConfirmStringsMock = ConfirmStringsMock;
});
define("xrm-mock/globalcontext/context.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ContextMock = /** @class */ (function () {
        function ContextMock(components) {
            this.userSettings = components.userSettings;
            this.client = components.clientContext;
            this.clientUrl = components.clientUrl;
            this.currentTheme = components.currentTheme;
            this.isAutoSaveEnabled = components.isAutoSaveEnabled;
            this.orgLcid = components.orgLcid;
            this.orgUniqueName = components.orgUniqueName;
            this.timeZoneOffset = components.timeZoneOffset;
            this.userId = components.userId;
            this.userLcid = components.userLcid;
            this.userName = components.userName;
            this.userRoles = components.userRoles;
            this.version = components.version;
        }
        ContextMock.prototype.getClientUrl = function () {
            return this.clientUrl;
        };
        ContextMock.prototype.getCurrentTheme = function () {
            return this.currentTheme;
        };
        ContextMock.prototype.getIsAutoSaveEnabled = function () {
            return this.isAutoSaveEnabled;
        };
        ContextMock.prototype.getOrgLcid = function () {
            return this.orgLcid;
        };
        ContextMock.prototype.getOrgUniqueName = function () {
            return this.orgUniqueName;
        };
        ContextMock.prototype.getQueryStringParameters = function () {
            throw new Error(("get query string parameters not implemented"));
        };
        ContextMock.prototype.getTimeZoneOffsetMinutes = function () {
            return this.timeZoneOffset;
        };
        ContextMock.prototype.getUserSettings = function () {
            return this.userSettings;
        };
        ContextMock.prototype.getUserId = function () {
            return this.userId;
        };
        ContextMock.prototype.getUserLcid = function () {
            return this.userLcid;
        };
        ContextMock.prototype.getUserName = function () {
            return this.userName;
        };
        ContextMock.prototype.getUserRoles = function () {
            return this.userRoles;
        };
        ContextMock.prototype.getVersion = function () {
            return this.version;
        };
        ContextMock.prototype.prependOrgName = function (sPath) {
            return sPath + this.orgUniqueName;
        };
        ContextMock.prototype.getAdvancedConfigSetting = function (setting) {
            throw new Error("Method not implemented.");
        };
        ContextMock.prototype.getCurrentAppName = function () {
            throw new Error("Method not implemented.");
        };
        ContextMock.prototype.getCurrentAppProperties = function () {
            throw new Error("Method not implemented.");
        };
        ContextMock.prototype.getCurrentAppUrl = function () {
            throw new Error("Method not implemented.");
        };
        ContextMock.prototype.isOnPremise = function () {
            throw new Error("Method not implemented.");
        };
        return ContextMock;
    }());
    exports.ContextMock = ContextMock;
});
define("xrm-mock/controls/controlnotificationaction/controlnotificationaction.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ControlNotificationActionMock = /** @class */ (function () {
        function ControlNotificationActionMock(actions, message) {
            this.message = message;
            this.actions = actions;
        }
        return ControlNotificationActionMock;
    }());
    exports.ControlNotificationActionMock = ControlNotificationActionMock;
});
define("xrm-mock/data/data.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DataMock = /** @class */ (function () {
        function DataMock(entity, process) {
            this.entity = entity;
            this.process = process;
        }
        DataMock.prototype.refresh = function (save) {
            throw new Error(("refresh not implemented"));
        };
        DataMock.prototype.save = function (saveOptions) {
            throw new Error(("save not implemented"));
        };
        return DataMock;
    }());
    exports.DataMock = DataMock;
});
define("xrm-mock/controls/datecontrol/datecontrol.mock", ["require", "exports", "xrm-mock/controls/standardcontrol/standardcontrol.mock"], function (require, exports, standardcontrol_mock_3) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DateControlMock = /** @class */ (function (_super) {
        __extends(DateControlMock, _super);
        function DateControlMock(components) {
            var _this = _super.call(this, DateControlMock.defaultComponents(components)) || this;
            _this.showTime = components.showTime;
            return _this;
        }
        DateControlMock.defaultComponents = function (components) {
            if (!("showTime" in components)) {
                components.showTime = components.attribute
                    ? components.attribute.getFormat() === "datetime"
                    : false;
            }
            return components;
        };
        DateControlMock.prototype.getShowTime = function () {
            return this.showTime;
        };
        DateControlMock.prototype.setShowTime = function (showTime) {
            // TODO: Does the Attribute format need to get updated as well?
            this.showTime = showTime;
        };
        return DateControlMock;
    }(standardcontrol_mock_3.StandardControlMock));
    exports.DateControlMock = DateControlMock;
});
define("xrm-mock/attributes/dateattribute/dateattribute.mock", ["require", "exports", "xrm-mock/attributes/attribute/attribute.mock"], function (require, exports, attribute_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DateAttributeMock = /** @class */ (function (_super) {
        __extends(DateAttributeMock, _super);
        function DateAttributeMock(components) {
            return _super.call(this, DateAttributeMock.defaultComponents(components)) || this;
        }
        DateAttributeMock.create = function (name, value) {
            return new DateAttributeMock({ name: name, value: value });
        };
        DateAttributeMock.defaultComponents = function (components) {
            if (!components.format) {
                components.format = "date";
            }
            components.attributeType = "datetime";
            return components;
        };
        DateAttributeMock.prototype.getFormat = function () {
            return _super.prototype.getFormat.call(this);
        };
        return DateAttributeMock;
    }(attribute_mock_2.AttributeMock));
    exports.DateAttributeMock = DateAttributeMock;
});
define("xrm-mock/dateformattinginfo/dateformattinginfo.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DateFormattingInfoMock = /** @class */ (function () {
        function DateFormattingInfoMock(components) {
            this.amDesignator = components.amDesignator;
            this.abbreviatedDayNames = components.abbreviatedDayNames;
            this.abbreviatedMonthGenitiveNames = components.abbreviatedMonthGenitiveNames;
            this.abbreviatedMonthNames = components.abbreviatedMonthNames;
            this.calendarWeekRule = components.calendarWeekRule;
            this.calendar = components.calendar;
            this.dateSeparator = components.dateSeparator;
            this.dayNames = components.dayNames;
            this.firstDayOfWeek = components.firstDayOfWeek;
            this.fullDateTimePattern = components.fullDateTimePattern;
            this.longDatePattern = components.longDatePattern;
            this.longTimePattern = components.longTimePattern;
            this.monthDayPattern = components.monthDayPattern;
            this.monthGenitiveNames = components.monthGenitiveNames;
            this.monthNames = components.monthNames;
            this.pmDesignator = components.pmDesignator;
            this.shortDatePattern = components.shortDatePattern;
            this.shortTimePattern = components.shortTimePattern;
            this.shortestDayNames = components.shortestDayNames;
            this.sortableDateTimePattern = components.sortableDateTimePattern;
            this.timeSeparator = components.timeSeparator;
            this.universalSortableDateTimePattern = components.universalSortableDateTimePattern;
            this.yearMonthPattern = components.yearMonthPattern;
        }
        return DateFormattingInfoMock;
    }());
    exports.DateFormattingInfoMock = DateFormattingInfoMock;
});
define("xrm-mock/device/device.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DeviceMock = /** @class */ (function () {
        function DeviceMock() {
        }
        DeviceMock.prototype.captureAudio = function () {
            throw new Error("Not implemented.");
        };
        DeviceMock.prototype.captureImage = function (imageOptions) {
            throw new Error("Not implemented.");
        };
        DeviceMock.prototype.captureVideo = function () {
            throw new Error("Not implemented.");
        };
        DeviceMock.prototype.getBarcodeValue = function () {
            throw new Error("Not implemented.");
        };
        DeviceMock.prototype.getCurrentPosition = function () {
            throw new Error("Not implemented.");
        };
        DeviceMock.prototype.pickFile = function (pickFileOptions) {
            throw new Error("Not implemented.");
        };
        return DeviceMock;
    }());
    exports.DeviceMock = DeviceMock;
});
define("xrm-mock/navigation/dialogsizeoptions/dialogsizeoptions.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DialogSizeOptionsMock = /** @class */ (function () {
        function DialogSizeOptionsMock(height, width) {
            this.height = height;
            this.width = width;
        }
        return DialogSizeOptionsMock;
    }());
    exports.DialogSizeOptionsMock = DialogSizeOptionsMock;
});
define("xrm-mock/encoding/encoding.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var EncodingMock = /** @class */ (function () {
        function EncodingMock() {
        }
        EncodingMock.prototype.xmlAttributeEncode = function (arg) {
            throw new Error("Not implemented");
        };
        EncodingMock.prototype.xmlEncode = function (arg) {
            throw new Error("Not implemented");
        };
        return EncodingMock;
    }());
    exports.EncodingMock = EncodingMock;
});
define("xrm-mock/navigation/entityformoptions/entityformoptions.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var EntityFormOptionsMock = /** @class */ (function () {
        function EntityFormOptionsMock(components) {
            this.cmdbar = components.cmdbar;
            this.createFromEntity = components.createFromEntity;
            this.entityId = components.entityId;
            this.entityName = components.entityName;
            this.formId = components.formId;
            this.height = components.height;
            this.isCrossEntityNavigate = components.isCrossEntityNavigate;
            this.isOfflineSyncError = components.isOfflineSyncError;
            this.navBar = components.navBar;
            this.openInNewWindow = components.openInNewWindow;
            this.windowPosition = components.windowPosition;
            this.processId = components.processId;
            this.processInstanceId = components.processInstanceId;
            this.relationship = components.relationship;
            this.selectedStageId = components.selectedStageId;
            this.useQuickCreateForm = components.useQuickCreateForm;
            this.width = components.width;
        }
        return EntityFormOptionsMock;
    }());
    exports.EntityFormOptionsMock = EntityFormOptionsMock;
});
define("xrm-mock/metadata/entitymetadata/entitymetadata.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var EntityMetadataMock = /** @class */ (function () {
        function EntityMetadataMock(components) {
            this.ActivityTypeMask = components.ActivityTypeMask;
            this.Attributes = components.Attributes;
            this.AutoRouteToOwnerQueue = components.AutoRouteToOwnerQueue;
            this.CanEnableSyncToExternalSearchIndex = components.CanEnableSyncToExternalSearchIndex;
            this.CanBeInManyToMany = components.CanBeInManyToMany;
            this.CanBePrimaryEntityInRelationship = components.CanBePrimaryEntityInRelationship;
            this.CanBeRelatedEntityInRelationship = components.CanBeRelatedEntityInRelationship;
            this.CanCreateAttributes = components.CanCreateAttributes;
            this.CanCreateCharts = components.CanCreateCharts;
            this.CanCreateForms = components.CanCreateForms;
            this.CanCreateViews = components.CanCreateViews;
            this.CanModifyAdditionalSettings = components.CanModifyAdditionalSettings;
            this.CanTriggerWorkflow = components.CanTriggerWorkflow;
            this.Description = components.Description;
            this.DisplayCollectionName = components.DisplayCollectionName;
            this.DisplayName = components.DisplayName;
            this.EntityColor = components.EntityColor;
            this.EntitySetName = components.EntitySetName;
            this.IconLargeName = components.IconLargeName;
            this.IconMediumName = components.IconMediumName;
            this.IconSmallName = components.IconSmallName;
            this.IsActivity = components.IsActivity;
            this.IsActivityParty = components.IsActivityParty;
            this.IsAuditEnabled = components.IsAuditEnabled;
            this.IsAvailableOffline = components.IsAvailableOffline;
            this.IsBPFEntity = components.IsBPFEntity;
            this.IsChildEntity = components.IsChildEntity;
            this.IsConnectionsEnabled = components.IsConnectionsEnabled;
            this.IsCustomEntity = components.IsCustomEntity;
            this.IsCustomizable = components.IsCustomizable;
            this.IsDocumentManagementEnabled = components.IsDocumentManagementEnabled;
            this.IsDuplicateDetectionEnabled = components.IsDuplicateDetectionEnabled;
            this.IsEnabledForCharts = components.IsEnabledForCharts;
            this.IsOneNotIntegrationEnabled = components.IsOneNotIntegrationEnabled;
            this.IsOptimisitcConcurrencyEnabled = components.IsOptimisitcConcurrencyEnabled;
            this.IsQuickCreateEnabled = components.IsQuickCreateEnabled;
            this.IsImportable = components.IsImportable;
            this.IsIntersect = components.IsIntersect;
            this.IsMailMergeEnabled = components.IsMailMergeEnabled;
            this.IsManaged = components.IsManaged;
            this.IsMappable = components.IsMappable;
            this.IsReadingPaneEnabled = components.IsReadingPaneEnabled;
            this.IsRenameable = components.IsRenameable;
            this.IsStateModelAware = components.IsStateModelAware;
            this.IsValidForAdvancedFind = components.IsValidForAdvancedFind;
            this.IsValidForQueue = components.IsValidForQueue;
            this.IsVisibleInMobileClient = components.IsVisibleInMobileClient;
            this.LogicalCollectionName = components.LogicalCollectionName;
            this.LogicalName = components.LogicalName;
            this.ObjectTypeCode = components.ObjectTypeCode;
            this.OwnershipTypeCode = components.OwnershipTypeCode;
            this.PrimaryIdAttribute = components.PrimaryIdAttribute;
            this.PrimaryNameAttribute = components.PrimaryNameAttribute;
            this.RecurrenceBaseEntityLogicalName = components.RecurrenceBaseEntityLogicalName;
            this.PrimaryImageAttribute = components.PrimaryImageAttribute;
        }
        return EntityMetadataMock;
    }());
    exports.EntityMetadataMock = EntityMetadataMock;
});
define("xrm-mock/entity/entity.mock", ["require", "exports", "xrm-mock/collection/itemcollection/itemcollection.mock"], function (require, exports, itemcollection_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var EntityMock = /** @class */ (function () {
        function EntityMock(components) {
            components = components || {};
            this.id = components.id || "{deadbeef-dead-beef-dead-beefdeadbeaf}";
            this.entityName = components.entityName || "contact";
            this.primaryValue = components.primaryValue || "Default Contact";
            this.attributes = components.attributes || new itemcollection_mock_2.ItemCollectionMock();
            this.saveEventHandlers = [];
        }
        EntityMock.prototype.addOnSave = function (handler) {
            this.saveEventHandlers.push(handler);
        };
        EntityMock.prototype.getEntityName = function () {
            return this.entityName;
        };
        EntityMock.prototype.getDataXml = function () {
            throw new Error(("getDataXml not implemented"));
        };
        EntityMock.prototype.getEntityReference = function () {
            return {
                entityType: this.entityName,
                id: this.id,
                name: this.primaryValue,
            };
        };
        EntityMock.prototype.getId = function () {
            return this.id;
        };
        EntityMock.prototype.getIsDirty = function () {
            var isDirty = false;
            if (this.attributes) {
                for (var i = 0; i < this.attributes.getLength(); i++) {
                    var attr = this.attributes.get(i);
                    if (attr.getIsDirty()) {
                        isDirty = true;
                        break;
                    }
                }
            }
            return isDirty;
        };
        EntityMock.prototype.getPrimaryAttributeValue = function () {
            return this.primaryValue;
        };
        EntityMock.prototype.isValid = function () {
            throw new Error("isValid not implemented.");
        };
        EntityMock.prototype.removeOnSave = function (handler) {
            var index = this.saveEventHandlers.indexOf(handler);
            this.saveEventHandlers.splice(index);
        };
        EntityMock.prototype.save = function (saveMode) {
            var context = this.getSaveContext(saveMode);
            var _loop_1 = function (handler) {
                var index = this_1.saveEventHandlers.indexOf(handler);
                context.getDepth = function () { return index; };
                handler(context);
            };
            var this_1 = this;
            for (var _i = 0, _a = this.saveEventHandlers; _i < _a.length; _i++) {
                var handler = _a[_i];
                _loop_1(handler);
            }
        };
        EntityMock.prototype.getSaveContext = function (saveMode) {
            var _this = this;
            return {
                getContext: function () {
                    throw new Error("getContext not implemented.");
                },
                getDepth: null,
                getEventArgs: function () {
                    return _this.getSaveEventArgs(saveMode);
                },
                getEventSource: function () {
                    throw new Error("getEventSource not implemented.");
                },
                getFormContext: function () {
                    throw new Error("getFormContext not implemented.");
                },
                getSharedVariable: function () {
                    throw new Error("getSharedVariable not implemented.");
                },
                setSharedVariable: function () {
                    throw new Error("setSharedVariable not implemented.");
                },
            };
        };
        EntityMock.prototype.getSaveEventArgs = function (saveMode) {
            return {
                getSaveMode: function () {
                    var mode;
                    if (saveMode == null) {
                        mode = 1 /* Save */;
                    }
                    else if (saveMode === "saveandclose") {
                        mode = 2 /* SaveAndClose */;
                    }
                    else if (saveMode === "saveandnew") {
                        mode = 59 /* SaveAndNew */;
                    }
                    return mode;
                },
                isDefaultPrevented: function () { return false; },
                preventDefault: function () {
                    throw new Error("preventDefault not implemented.");
                },
            };
        };
        return EntityMock;
    }());
    exports.EntityMock = EntityMock;
});
define("xrm-mock/navigation/errordialogoptions/errordialogoptions.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ErrorDialogOptionsMock = /** @class */ (function () {
        function ErrorDialogOptionsMock(details, errorCode, message) {
            this.details = details;
            this.errorCode = errorCode;
            this.message = message;
        }
        return ErrorDialogOptionsMock;
    }());
    exports.ErrorDialogOptionsMock = ErrorDialogOptionsMock;
});
define("xrm-mock/errorresponse/errorresponse.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ErrorResponseMock = /** @class */ (function () {
        function ErrorResponseMock(errorCode, message) {
            this.errorCode = errorCode;
            this.message = message;
        }
        return ErrorResponseMock;
    }());
    exports.ErrorResponseMock = ErrorResponseMock;
});
define("xrm-mock/events/eventcontext/eventcontext.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var EventContextMock = /** @class */ (function () {
        function EventContextMock(components) {
            this.formContext = components.formContext;
            this.context = components.context;
        }
        EventContextMock.prototype.getContext = function () {
            return this.context;
        };
        EventContextMock.prototype.getDepth = function () {
            throw new Error("not implemented");
        };
        EventContextMock.prototype.getEventArgs = function () {
            throw new Error("not implemented");
        };
        EventContextMock.prototype.getEventSource = function () {
            throw new Error("not implemented");
        };
        EventContextMock.prototype.getFormContext = function () {
            return this.formContext;
        };
        EventContextMock.prototype.getSharedVariable = function (key) {
            throw new Error("not implemented");
        };
        EventContextMock.prototype.setSharedVariable = function (key, value) {
            throw new Error("not implemented");
        };
        return EventContextMock;
    }());
    exports.EventContextMock = EventContextMock;
});
define("xrm-mock/executeresponse/executeresponse.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ExecuteResponseMock = /** @class */ (function () {
        function ExecuteResponseMock(components) {
            this.body = components.body;
            this.headers = components.headers;
            this.ok = components.ok;
            this.status = components.status;
            this.statusText = components.statusText;
            this.type = components.type;
            this.url = components.url;
        }
        return ExecuteResponseMock;
    }());
    exports.ExecuteResponseMock = ExecuteResponseMock;
});
define("xrm-mock/navigation/filedetails/filedetails.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FileDetailsMock = /** @class */ (function () {
        function FileDetailsMock(fileContent, fileName, fileSize, mimeType) {
            this.fileContent = fileContent;
            this.fileName = fileName;
            this.fileSize = fileSize;
            this.mimeType = mimeType;
        }
        return FileDetailsMock;
    }());
    exports.FileDetailsMock = FileDetailsMock;
});
define("xrm-mock/formcontext/formcontext.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FormContextMock = /** @class */ (function () {
        function FormContextMock(data, ui) {
            this.data = data;
            this.ui = ui;
        }
        FormContextMock.prototype.getAttribute = function (param) {
            if (!arguments.length) {
                return this.data.entity.attributes.get();
            }
            else if (param && typeof param === "string") {
                return this.data.entity.attributes.get(param);
            }
            else if (typeof param === "number") {
                return this.data.entity.attributes.get(param);
            }
            else if (typeof param === "function") {
                return this.data.entity.attributes.get(param);
            }
            throw new Error("Collection.Get called with unknown parameter type: " + typeof param);
        };
        FormContextMock.prototype.getControl = function (param) {
            if (!arguments.length || param === undefined || param === null) {
                return this.ui.controls.get();
            }
            else if (typeof param === "string") {
                return this.ui.controls.get(param);
            }
            else if (typeof param === "number") {
                return this.ui.controls.get(param);
            }
            else if (typeof param === "function") {
                return this.ui.controls.get(param);
            }
            throw new Error("Collection.Get called with unknown parameter type: " + typeof param);
        };
        return FormContextMock;
    }());
    exports.FormContextMock = FormContextMock;
});
define("xrm-mock/controls/formitem/formitem.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FormItemMock = /** @class */ (function () {
        function FormItemMock(components) {
            this.id = components.id;
            this.label = components.label;
            this.formType = components.formType;
            this.currentItem = components.currentItem;
        }
        FormItemMock.prototype.getId = function () {
            return this.id;
        };
        FormItemMock.prototype.getLabel = function () {
            return this.label;
        };
        FormItemMock.prototype.navigate = function () {
            throw new Error("Form navigation not implemented.");
        };
        return FormItemMock;
    }());
    exports.FormItemMock = FormItemMock;
});
define("xrm-mock/utility/formopenparameters/formopenparameters.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FormOpenParametersMock = /** @class */ (function () {
        function FormOpenParametersMock(formid, navbar, cmdbar) {
            this.formid = formid;
            this.navbar = navbar;
            this.cmdbar = cmdbar;
        }
        return FormOpenParametersMock;
    }());
    exports.FormOpenParametersMock = FormOpenParametersMock;
});
define("xrm-mock/controls/formselector/formselector.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FormSelectorMock = /** @class */ (function () {
        function FormSelectorMock(items) {
            this.items = items;
        }
        FormSelectorMock.prototype.getCurrentItem = function () {
            var currentItem;
            if (this.items.itemCollection && this.items.itemCollection.length) {
                currentItem = this.items.itemCollection.filter(function (i) { return i.currentItem; })[0] || null;
            }
            return currentItem;
        };
        return FormSelectorMock;
    }());
    exports.FormSelectorMock = FormSelectorMock;
});
define("xrm-mock/device/getcurrentpositionresponse/getcurrentpositionresponse.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GetCurrentPositionResponseMock = /** @class */ (function () {
        function GetCurrentPositionResponseMock(coords, timestamp) {
            this.coords = coords;
            this.timestamp = timestamp;
        }
        return GetCurrentPositionResponseMock;
    }());
    exports.GetCurrentPositionResponseMock = GetCurrentPositionResponseMock;
});
define("xrm-mock/controls/gridcontrol/gridcontrol.mock", ["require", "exports", "xrm-mock/controls/control/control.mock"], function (require, exports, control_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GridControlMock = /** @class */ (function (_super) {
        __extends(GridControlMock, _super);
        function GridControlMock(components) {
            var _this = _super.call(this, GridControlMock.defaultComponents(components)) || this;
            _this.contextType = components.contextType;
            _this.entityName = components.entityName;
            _this.onLoadHandlers = components.onLoadHandlers;
            return _this;
        }
        GridControlMock.defaultComponents = function (components) {
            components.controlType = "subgrid";
            components.contextType = components.contextType || 4;
            components.entityName = components.entityName || "contact";
            components.onLoadHandlers = components.onLoadHandlers || [];
            return components;
        };
        GridControlMock.prototype.addOnLoad = function (handler) {
            this.onLoadHandlers.push(handler);
        };
        GridControlMock.prototype.getContextType = function () {
            return this.contextType;
        };
        GridControlMock.prototype.getEntityName = function () {
            return this.entityName;
        };
        GridControlMock.prototype.getGrid = function () {
            throw new Error("Method not implemented.");
        };
        GridControlMock.prototype.getViewSelector = function () {
            throw new Error("Method not implemented.");
        };
        GridControlMock.prototype.refresh = function () {
            throw new Error("Method not implemented.");
        };
        GridControlMock.prototype.removeOnLoad = function (handler) {
            for (var i = this.onLoadHandlers.length - 1; i >= 0; i--) {
                var func = this.onLoadHandlers[i];
                if (func === handler) {
                    this.onLoadHandlers.splice(i, 1);
                    break;
                }
            }
        };
        return GridControlMock;
    }(control_mock_2.ControlMock));
    exports.GridControlMock = GridControlMock;
});
define("xrm-mock/controls/grid/gridentity/gridentity.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GridEntityMock = /** @class */ (function () {
        function GridEntityMock(reference) {
            this.reference = reference;
        }
        GridEntityMock.prototype.getEntityName = function () {
            return this.reference.entityType;
        };
        GridEntityMock.prototype.getEntityReference = function () {
            return this.reference;
        };
        GridEntityMock.prototype.getId = function () {
            return this.reference.id;
        };
        GridEntityMock.prototype.getPrimaryAttributeValue = function () {
            return this.reference.name;
        };
        return GridEntityMock;
    }());
    exports.GridEntityMock = GridEntityMock;
});
define("xrm-mock/controls/grid/grid.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GridMock = /** @class */ (function () {
        function GridMock(rows) {
            this.rows = rows;
        }
        GridMock.prototype.getRows = function () {
            return this.rows;
        };
        GridMock.prototype.getSelectedRows = function () {
            throw new Error("Not implemented.");
        };
        GridMock.prototype.getTotalRecordCount = function () {
            throw new Error("Not implemented.");
        };
        return GridMock;
    }());
    exports.GridMock = GridMock;
});
define("xrm-mock/controls/grid/gridrow/gridrow.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GridRowMock = /** @class */ (function () {
        function GridRowMock(data, gridRowData) {
            this.data = data;
            this.gridRowData = gridRowData;
        }
        GridRowMock.prototype.getData = function () {
            return this.gridRowData;
        };
        return GridRowMock;
    }());
    exports.GridRowMock = GridRowMock;
});
define("xrm-mock/controls/grid/gridrowdata/gridrowdata.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GridRowDataMock = /** @class */ (function () {
        function GridRowDataMock(entity) {
            this.entity = entity;
        }
        GridRowDataMock.prototype.getEntity = function () {
            return this.entity;
        };
        return GridRowDataMock;
    }());
    exports.GridRowDataMock = GridRowDataMock;
});
define("xrm-mock/metadata/label/label.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var LabelMock = /** @class */ (function () {
        function LabelMock(localLabels, userLocalLabels) {
            this.LocalizedLabels = localLabels;
            this.UserLocalizedLabel = userLocalLabels;
        }
        return LabelMock;
    }());
    exports.LabelMock = LabelMock;
});
define("xrm-mock/metadata/localizedlabel/localizedlabel.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var LocalizedLabelMock = /** @class */ (function () {
        function LocalizedLabelMock(label, languageCode) {
            this.Label = label;
            this.LanguageCode = languageCode;
        }
        return LocalizedLabelMock;
    }());
    exports.LocalizedLabelMock = LocalizedLabelMock;
});
define("xrm-mock/controls/lookupcontrol/lookupcontrol.mock", ["require", "exports", "xrm-mock/controls/standardcontrol/standardcontrol.mock"], function (require, exports, standardcontrol_mock_4) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var LookupControlMock = /** @class */ (function (_super) {
        __extends(LookupControlMock, _super);
        function LookupControlMock(components) {
            var _this = _super.call(this, LookupControlMock.defaultComponents(components)) || this;
            _this.preSearchHandlers = [];
            _this.views = components.views || [];
            _this.filters = components.filters || [];
            _this.preSearchHandlers = components.preSearchHandlers || [];
            if (_this.views && _this.views.length > 1) {
                var defaultViews = _this.views.filter(function (v) { return v.isDefault; }).length;
                if (defaultViews > 1) {
                    throw new Error("Lookup Control cannot have more than one default view.");
                }
                else if (defaultViews === 0) {
                    _this.views[0].isDefault = true;
                }
            }
            return _this;
        }
        LookupControlMock.defaultComponents = function (components) {
            components.controlType = "lookup";
            return components;
        };
        LookupControlMock.prototype.addPreSearch = function (handler) {
            this.preSearchHandlers.push(handler);
        };
        LookupControlMock.prototype.addCustomFilter = function (filter, entityLogicalName) {
            this.filters.push({ filter: filter, entityLogicalName: entityLogicalName });
        };
        LookupControlMock.prototype.addCustomView = function (viewId, entityName, viewDisplayName, fetchXml, layoutXml, isDefault) {
            if (isDefault && this.getDefaultView()) {
                throw new Error("Lookup Control cannot have more than one default view.");
            }
            this.views.push({
                entityName: entityName,
                fetchXml: fetchXml,
                isDefault: isDefault,
                layoutXml: layoutXml,
                viewDisplayName: viewDisplayName,
                viewId: viewId,
            });
        };
        LookupControlMock.prototype.getDefaultView = function () {
            for (var _i = 0, _a = this.views; _i < _a.length; _i++) {
                var view = _a[_i];
                if (view.isDefault) {
                    return view.viewId;
                }
            }
            throw new Error("No default view was found!");
        };
        LookupControlMock.prototype.getEntityTypes = function () {
            throw new Error("Method not implemented.");
        };
        LookupControlMock.prototype.setEntityTypes = function (entityLogicalNames) {
            throw new Error("Method not implemented.");
        };
        LookupControlMock.prototype.removePreSearch = function (handler) {
            throw new Error("remove presearch not implemented");
        };
        LookupControlMock.prototype.setDefaultView = function (viewGuid) {
            for (var _i = 0, _a = this.views; _i < _a.length; _i++) {
                var view = _a[_i];
                view.isDefault = view.viewId === viewGuid;
            }
        };
        return LookupControlMock;
    }(standardcontrol_mock_4.StandardControlMock));
    exports.LookupControlMock = LookupControlMock;
});
define("xrm-mock/attributes/lookupattribute/lookupattribute.mock", ["require", "exports", "xrm-mock/attributes/attribute/attribute.mock"], function (require, exports, attribute_mock_3) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var LookupAttributeMock = /** @class */ (function (_super) {
        __extends(LookupAttributeMock, _super);
        function LookupAttributeMock(components) {
            var _this = _super.call(this, LookupAttributeMock.defaultComponents(components)) || this;
            _this.isPartyList = components.isPartyList;
            return _this;
        }
        LookupAttributeMock.defaultComponents = function (components) {
            if (!("isPartyList" in components)) {
                components.isPartyList = false;
            }
            components.attributeType = "lookup";
            return components;
        };
        LookupAttributeMock.prototype.getIsPartyList = function () {
            return this.isPartyList;
        };
        return LookupAttributeMock;
    }(attribute_mock_3.AttributeMock));
    exports.LookupAttributeMock = LookupAttributeMock;
});
define("xrm-mock/lookupoptions/lookupoptions.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var LookupOptionsMock = /** @class */ (function () {
        function LookupOptionsMock(components) {
            this.allowMultiSelect = components.allowMultiSelect;
            this.defaultEntityType = components.defaultEntityType;
            this.defaultViewId = components.defaultViewId;
            this.entityTypes = components.entityTypes;
            this.showBarcodeScanner = components.showBarcodeScanner;
            this.viewIds = components.viewIds;
        }
        return LookupOptionsMock;
    }());
    exports.LookupOptionsMock = LookupOptionsMock;
});
define("xrm-mock/lookupvalue/lookupvalue.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var LookupValueMock = /** @class */ (function () {
        function LookupValueMock(id, entityType, name) {
            this.id = id;
            this.entityType = entityType;
            this.name = name;
        }
        return LookupValueMock;
    }());
    exports.LookupValueMock = LookupValueMock;
});
define("xrm-mock/controls/navigation/navigation.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var NavigationMock = /** @class */ (function () {
        function NavigationMock(items) {
            this.items = items;
        }
        return NavigationMock;
    }());
    exports.NavigationMock = NavigationMock;
});
define("xrm-mock/controls/navigationitem/navigationitem.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var NavigationItemMock = /** @class */ (function () {
        function NavigationItemMock(id, standardElement, focusable) {
            this.id = id;
            this.standardElement = standardElement;
            this.focusable = focusable;
        }
        NavigationItemMock.prototype.setVisible = function (visible) {
            return this.standardElement.setVisible(visible);
        };
        NavigationItemMock.prototype.getVisible = function () {
            return this.standardElement.getVisible();
        };
        NavigationItemMock.prototype.getLabel = function () {
            return this.standardElement.getLabel();
        };
        NavigationItemMock.prototype.setLabel = function (label) {
            return this.standardElement.setLabel(label);
        };
        NavigationItemMock.prototype.setFocus = function () {
            return this.focusable.setFocus();
        };
        NavigationItemMock.prototype.getId = function () {
            return this.id;
        };
        return NavigationItemMock;
    }());
    exports.NavigationItemMock = NavigationItemMock;
});
define("xrm-mock/navigation/navigation.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var NavigationStaticMock = /** @class */ (function () {
        function NavigationStaticMock() {
            this.notImplementedError = "Navigation methods not implemented. Consider stubbing calls using a tool such as Sinon.JS";
        }
        NavigationStaticMock.prototype.openAlertDialog = function (alertStrings, alertOptions) {
            throw new Error(this.notImplementedError);
        };
        NavigationStaticMock.prototype.openConfirmDialog = function (param) {
            throw new Error(this.notImplementedError);
        };
        NavigationStaticMock.prototype.openErrorDialog = function (errorOptions) {
            throw new Error(this.notImplementedError);
        };
        NavigationStaticMock.prototype.openFile = function (file, openFileOptions) {
            throw new Error(this.notImplementedError);
        };
        NavigationStaticMock.prototype.openForm = function (entityFormOptions, formParameters) {
            throw new Error(this.notImplementedError);
        };
        NavigationStaticMock.prototype.openUrl = function (url, openUrlOptions) {
            throw new Error(this.notImplementedError);
        };
        NavigationStaticMock.prototype.openWebResource = function (webResourceName, windowOptions, data) {
            throw new Error(this.notImplementedError);
        };
        return NavigationStaticMock;
    }());
    exports.NavigationStaticMock = NavigationStaticMock;
});
define("xrm-mock/controls/numbercontrol/numbercontrol.mock", ["require", "exports", "xrm-mock/controls/autolookupcontrol/autolookupcontrol.mock"], function (require, exports, autolookupcontrol_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var NumberControlMock = /** @class */ (function (_super) {
        __extends(NumberControlMock, _super);
        function NumberControlMock(components) {
            return _super.call(this, NumberControlMock.defaultComponents(components)) || this;
        }
        NumberControlMock.defaultComponents = function (components) {
            return components;
        };
        return NumberControlMock;
    }(autolookupcontrol_mock_1.AutoLookupControlMock));
    exports.NumberControlMock = NumberControlMock;
});
define("xrm-mock/attributes/numberattribute/numberattribute.mock", ["require", "exports", "xrm-mock/attributes/attribute/attribute.mock"], function (require, exports, attribute_mock_4) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var NumberAttributeMock = /** @class */ (function (_super) {
        __extends(NumberAttributeMock, _super);
        function NumberAttributeMock(components) {
            var _this = _super.call(this, NumberAttributeMock.defaultComponents(components)) || this;
            _this.min = components.min;
            _this.max = components.max;
            _this.precision = components.precision;
            _this.validatePrecision();
            return _this;
        }
        NumberAttributeMock.defaultComponents = function (components) {
            components.format = components.format || "none";
            components.precision = components.precision || 0;
            if (!components.attributeType) {
                components.attributeType = "decimal";
            }
            return components;
        };
        NumberAttributeMock.prototype.getFormat = function () {
            return _super.prototype.getFormat.call(this);
        };
        NumberAttributeMock.prototype.getMax = function () {
            return this.max;
        };
        NumberAttributeMock.prototype.getMin = function () {
            return this.min;
        };
        NumberAttributeMock.prototype.getPrecision = function () {
            return this.precision;
        };
        NumberAttributeMock.prototype.setValue = function (value) {
            if ((this.min || this.min === 0) && this.min > value) {
                throw new Error(("value cannot be below the min of " + this.min));
            }
            else if ((this.max || this.max === 0) && this.max < value) {
                throw new Error(("value cannot be above the max of " + this.max));
            }
            else {
                _super.prototype.setValue.call(this, value);
            }
        };
        NumberAttributeMock.prototype.validatePrecision = function () {
            if (this.precision > 10) {
                throw new Error(("precision cannot be greater than 10, but was " + this.precision));
            }
            else if (this.precision < 0) {
                throw new Error(("precision cannot be less than 0, but was " + this.precision));
            }
        };
        return NumberAttributeMock;
    }(attribute_mock_4.AttributeMock));
    exports.NumberAttributeMock = NumberAttributeMock;
});
define("xrm-mock/utility/openparameters/openparameters.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OpenParametersMock = /** @class */ (function () {
        function OpenParametersMock() {
        }
        return OpenParametersMock;
    }());
    exports.OpenParametersMock = OpenParametersMock;
});
define("xrm-mock/navigation/openwebresourceoptions/openwebresourceoptions.mock", ["require", "exports", "xrm-mock/navigation/dialogsizeoptions/dialogsizeoptions.mock"], function (require, exports, dialogsizeoptions_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OpenWebresourceOptionsMock = /** @class */ (function (_super) {
        __extends(OpenWebresourceOptionsMock, _super);
        function OpenWebresourceOptionsMock(openInNewWindow, height, width) {
            var _this = _super.call(this, height, width) || this;
            _this.openInNewWindow = openInNewWindow;
            return _this;
        }
        return OpenWebresourceOptionsMock;
    }(dialogsizeoptions_mock_1.DialogSizeOptionsMock));
    exports.OpenWebresourceOptionsMock = OpenWebresourceOptionsMock;
});
define("xrm-mock/optionsetvalue/optionsetvalue.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OptionSetValueMock = /** @class */ (function () {
        function OptionSetValueMock(text, value) {
            this.text = text;
            this.value = value;
        }
        return OptionSetValueMock;
    }());
    exports.OptionSetValueMock = OptionSetValueMock;
});
define("xrm-mock/controls/optionsetcontrol/optionsetcontrol.mock", ["require", "exports", "xrm-mock/controls/standardcontrol/standardcontrol.mock"], function (require, exports, standardcontrol_mock_5) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OptionSetControlMock = /** @class */ (function (_super) {
        __extends(OptionSetControlMock, _super);
        function OptionSetControlMock(components) {
            var _this = _super.call(this, OptionSetControlMock.defaultComponents(components)) || this;
            _this.options = components.options || [];
            return _this;
        }
        OptionSetControlMock.defaultComponents = function (components) {
            components.controlType = "optionset";
            return components;
        };
        OptionSetControlMock.prototype.addOption = function (option, index) {
            index = index || index === 0
                ? index
                : this.options.length;
            this.options.splice(index, 0, option);
        };
        OptionSetControlMock.prototype.clearOptions = function () {
            var length = this.options.length;
            if (length) {
                this.options.splice(0, length);
            }
        };
        OptionSetControlMock.prototype.removeOption = function (value) {
            var option;
            for (var _i = 0, _a = this.options; _i < _a.length; _i++) {
                var item = _a[_i];
                if (item.value === value) {
                    option = item;
                    break;
                }
            }
            var index = this.options.indexOf(option);
            this.options.splice(index, 1);
        };
        return OptionSetControlMock;
    }(standardcontrol_mock_5.StandardControlMock));
    exports.OptionSetControlMock = OptionSetControlMock;
});
define("xrm-mock/attributes/optionsetattribute/optionsetattribute.mock", ["require", "exports", "xrm-mock/optionsetvalue/optionsetvalue.mock", "xrm-mock/attributes/enumattribute/enumattribute.mock"], function (require, exports, optionsetvalue_mock_1, enumattribute_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OptionSetAttributeMock = /** @class */ (function (_super) {
        __extends(OptionSetAttributeMock, _super);
        function OptionSetAttributeMock(components) {
            var _this = _super.call(this, OptionSetAttributeMock.defaultComponents(components)) || this;
            _this.options = components.options
                ? components.options.map(function (o) {
                    return new optionsetvalue_mock_1.OptionSetValueMock(o.text, o.value);
                })
                : components.value || components.value === 0
                    ? [{
                            text: components.value.toString(),
                            value: components.value,
                        }]
                    : [];
            _this.format = components.format || "language";
            _this.selectedOption = _this.getOption(_this.getValue());
            return _this;
        }
        OptionSetAttributeMock.create = function (name, value) {
            return new OptionSetAttributeMock({ name: name, value: value });
        };
        OptionSetAttributeMock.defaultComponents = function (components) {
            components.attributeType = "optionset";
            return components;
        };
        OptionSetAttributeMock.prototype.getFormat = function () {
            return _super.prototype.getFormat.call(this);
        };
        OptionSetAttributeMock.prototype.getOption = function (param) {
            return typeof param === "number"
                ? this.options.filter(function (o) { return o.value === param; })[0]
                : this.options.filter(function (o) { return o.text === param; })[0];
        };
        OptionSetAttributeMock.prototype.getOptions = function () {
            return this.options;
        };
        OptionSetAttributeMock.prototype.getSelectedOption = function () {
            return this.selectedOption;
        };
        OptionSetAttributeMock.prototype.getText = function () {
            return this.selectedOption.text;
        };
        OptionSetAttributeMock.prototype.setValue = function (value) {
            _super.prototype.setValue.call(this, value);
            this.selectedOption = this.getOption(value);
        };
        return OptionSetAttributeMock;
    }(enumattribute_mock_2.EnumAttributeMock));
    exports.OptionSetAttributeMock = OptionSetAttributeMock;
});
define("xrm-mock/metadata/optionmetadata/optionmetadata.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OptionMetadataMock = /** @class */ (function () {
        function OptionMetadataMock(components) {
            this.Value = components.Value;
            this.Color = components.Color;
            this.IsManaged = components.IsManaged;
            this.ExternalValue = components.ExternalValue;
            this.MetadataId = components.MetadataId;
            this.HasChanged = components.HasChanged;
            this.State = components.State;
            this.TransitionData = components.TransitionData;
            this.Label = components.Label;
            this.Description = components.Description;
        }
        return OptionMetadataMock;
    }());
    exports.OptionMetadataMock = OptionMetadataMock;
});
define("xrm-mock/organizationsettings/organizationsettings.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OrganizationSettingsMock = /** @class */ (function () {
        function OrganizationSettingsMock(components) {
            this.baseCurrencyId = components.baseCurrencyId;
            this.defaultCountryCode = components.defaultCountryCode;
            this.isAutoSaveEnabled = components.isAutoSaveEnabled;
            this.languageId = components.languageId;
            this.organizationId = components.organizationId;
            this.uniqueName = components.uniqueName;
            this.useSkypeProtocol = components.useSkypeProtocol;
        }
        return OrganizationSettingsMock;
    }());
    exports.OrganizationSettingsMock = OrganizationSettingsMock;
});
define("xrm-mock/page/page.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PageMock = /** @class */ (function () {
        function PageMock(context, formContext) {
            this.context = context;
            this.data = formContext.data;
            this.ui = formContext.ui;
        }
        PageMock.prototype.getAttribute = function (param) {
            if (!arguments.length) {
                return this.data.entity.attributes.get();
            }
            else if (param && typeof param === "string") {
                return this.data.entity.attributes.get(param);
            }
            else if (typeof param === "number") {
                return this.data.entity.attributes.get(param);
            }
            else if (typeof param === "function") {
                return this.data.entity.attributes.get(param);
            }
            throw new Error("Collection.Get called with unknown parameter type: " + typeof param);
        };
        PageMock.prototype.getControl = function (param) {
            if (!arguments.length || param === undefined || param === null) {
                return this.ui.controls.get();
            }
            else if (typeof param === "string") {
                return this.ui.controls.get(param);
            }
            else if (typeof param === "number") {
                return this.ui.controls.get(param);
            }
            else if (typeof param === "function") {
                return this.ui.controls.get(param);
            }
            throw new Error("Collection.Get called with unknown parameter type: " + typeof param);
        };
        return PageMock;
    }());
    exports.PageMock = PageMock;
});
define("xrm-mock/panel/panel.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PanelMock = /** @class */ (function () {
        function PanelMock() {
        }
        PanelMock.prototype.loadPanel = function (url, title) {
            throw new Error("Not implemented.");
        };
        return PanelMock;
    }());
    exports.PanelMock = PanelMock;
});
define("xrm-mock/device/pickfileoptions/pickfileoptions.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PickFileOptionsMock = /** @class */ (function () {
        function PickFileOptionsMock(accept, allowMultipleFiles, maximumAllowedFileSize) {
            this.accept = accept;
            this.allowMultipleFiles = allowMultipleFiles;
            this.maximumAllowedFileSize = maximumAllowedFileSize;
        }
        return PickFileOptionsMock;
    }());
    exports.PickFileOptionsMock = PickFileOptionsMock;
});
define("xrm-mock/privilege/privilege.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PrivilegeMock = /** @class */ (function () {
        function PrivilegeMock(canRead, canUpdate, canCreate) {
            this.canRead = canRead;
            this.canUpdate = canUpdate;
            this.canCreate = canCreate;
        }
        return PrivilegeMock;
    }());
    exports.PrivilegeMock = PrivilegeMock;
});
define("xrm-mock/controls/processcontrol/processcontrol.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ProcessControlMock = /** @class */ (function () {
        function ProcessControlMock(displayState, getVisibleElement, setVisibleElement) {
            this.displayState = displayState;
            this.getVisibleElement = getVisibleElement;
            this.setVisibleElement = setVisibleElement;
        }
        ProcessControlMock.prototype.getVisible = function () {
            return this.getVisibleElement.getVisible();
        };
        ProcessControlMock.prototype.setVisible = function (visible) {
            return this.setVisibleElement.setVisible(visible);
        };
        ProcessControlMock.prototype.setDisplayState = function (displayState) {
            this.displayState = displayState;
        };
        ProcessControlMock.prototype.getDisplayState = function () {
            return this.displayState;
        };
        ProcessControlMock.prototype.reflow = function (updateUI, parentStage, nextStage) {
            throw new Error("Not implemented.");
        };
        return ProcessControlMock;
    }());
    exports.ProcessControlMock = ProcessControlMock;
});
define("xrm-mock/processflow/processdictionary/processdictionary.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ProcessDictionaryMock = /** @class */ (function () {
        function ProcessDictionaryMock(index) {
            this[index] = index;
        }
        return ProcessDictionaryMock;
    }());
    exports.ProcessDictionaryMock = ProcessDictionaryMock;
});
define("xrm-mock/processflow/stage/stage.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StageMock = /** @class */ (function () {
        function StageMock(id, name, status, category, steps) {
            this.id = id;
            this.name = name;
            this.status = status;
            this.category = category;
            this.steps = steps;
        }
        StageMock.prototype.getCategory = function () {
            return {
                getValue: function () {
                    throw new Error("get category value not implemented");
                },
            };
        };
        StageMock.prototype.getEntityName = function () {
            throw new Error("get entity name not implemented");
        };
        StageMock.prototype.getId = function () {
            return this.id;
        };
        StageMock.prototype.getName = function () {
            return this.name;
        };
        StageMock.prototype.getStatus = function () {
            return this.status;
        };
        StageMock.prototype.getSteps = function () {
            return this.steps;
        };
        StageMock.prototype._setStatus = function (status) {
            this.status = status;
        };
        return StageMock;
    }());
    exports.StageMock = StageMock;
});
define("xrm-mock/processflow/processmanager/processmanager.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ProcessManagerMock = /** @class */ (function () {
        function ProcessManagerMock(enabledProcesses) {
            this.enabledProcesses = enabledProcesses;
        }
        ProcessManagerMock.prototype.getActiveProcess = function () {
            return this.enabledProcesses.filter(function (p) {
                return p.isRendered();
            })[0] || undefined;
        };
        ProcessManagerMock.prototype.setActiveProcess = function (processId, callbackFunction) {
            var activeProcess = this.getActiveProcess();
            if (activeProcess) {
                var newProcess = this.enabledProcesses.filter(function (p) {
                    return p.getId() === processId;
                })[0] || undefined;
                if (newProcess) {
                    activeProcess.isRendered = function () { return false; };
                    newProcess.isRendered = function () { return true; };
                }
            }
        };
        ProcessManagerMock.prototype.setActiveProcessInstance = function (processInstanceId, callbackFunction) {
            throw new Error("set active process instance not implemented.");
        };
        ProcessManagerMock.prototype.getProcessInstances = function (callbackFunction) {
            throw new Error("get process instances not implemented.");
        };
        ProcessManagerMock.prototype.addOnProcessStatusChange = function (handler) {
            throw new Error("add on process status change not implemented.");
        };
        ProcessManagerMock.prototype.removeOnProcessStatusChange = function (handler) {
            throw new Error("remove on process status change not implemented.");
        };
        ProcessManagerMock.prototype.getInstanceId = function () {
            var activeProcess = this.getActiveProcess();
            var instanceId = activeProcess.getId() || "";
            return instanceId;
        };
        ProcessManagerMock.prototype.getInstanceName = function () {
            var activeProcess = this.getActiveProcess();
            var instanceName = activeProcess.getName();
            return instanceName;
        };
        ProcessManagerMock.prototype.getStatus = function () {
            throw new Error("get status not implemented.");
        };
        ProcessManagerMock.prototype.setStatus = function (status, callbackFunction) {
            throw new Error("set status not implemented.");
        };
        ProcessManagerMock.prototype.getActiveStage = function () {
            var activeProcess = this.getActiveProcess();
            var stages = activeProcess.getStages();
            var activeStages = [];
            stages.forEach(function (s) {
                if (s.getStatus() === "active") {
                    activeStages.push(s);
                }
            });
            return activeStages[0] || undefined;
        };
        ProcessManagerMock.prototype.setActiveStage = function (stageId, callbackFunction) {
            var activeProcess = this.getActiveProcess();
            var stages = activeProcess.getStages();
            var setStatus = "";
            stages.forEach(function (s) {
                if (s.getId() === stageId) {
                    s._setStatus("active");
                    setStatus = "active";
                }
                else {
                    s._setStatus("inactive");
                }
            });
            if (callbackFunction) {
                callbackFunction(setStatus);
            }
        };
        ProcessManagerMock.prototype.getActivePath = function () {
            throw new Error("get active path not implemented");
        };
        ProcessManagerMock.prototype.getEnabledProcesses = function (callbackFunction) {
            callbackFunction(this.enabledProcesses);
        };
        ProcessManagerMock.prototype.getSelectedStage = function () {
            throw new Error("get selected not implemented");
        };
        ProcessManagerMock.prototype.addOnStageChange = function (handler) {
            throw new Error("add on stage change not implemented");
        };
        ProcessManagerMock.prototype.addOnStageSelected = function (handler) {
            throw new Error("add on stage selected not implemented");
        };
        ProcessManagerMock.prototype.removeOnStageChange = function (handler) {
            throw new Error("remove on stage change not implemented");
        };
        ProcessManagerMock.prototype.removeOnStageSelected = function (handler) {
            throw new Error("remove on stage selected not implemented");
        };
        ProcessManagerMock.prototype.moveNext = function (callbackFunction) {
            throw new Error("move next not implemented");
        };
        ProcessManagerMock.prototype.movePrevious = function (callbackFunction) {
            throw new Error("move previous not implemented");
        };
        return ProcessManagerMock;
    }());
    exports.ProcessManagerMock = ProcessManagerMock;
});
define("xrm-mock/processflow/process/process.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ProcessMock = /** @class */ (function () {
        function ProcessMock(components) {
            this.id = components.id;
            this.name = components.name;
            this.stages = components.stages;
            this.rendered = components.rendered;
        }
        ProcessMock.prototype.getId = function () {
            return this.id;
        };
        ProcessMock.prototype.getName = function () {
            return this.name;
        };
        ProcessMock.prototype.getStages = function () {
            return this.stages;
        };
        ProcessMock.prototype.isRendered = function () {
            return this.rendered;
        };
        return ProcessMock;
    }());
    exports.ProcessMock = ProcessMock;
});
define("xrm-mock/navigation/relationship/relationship.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var RelationshipMock = /** @class */ (function () {
        function RelationshipMock(components) {
            this.attributeName = components.attributeName;
            this.name = components.name;
            this.navigationPropertyName = components.navigationPropertyName;
            this.relationshipType = components.relationshipType;
            this.roleType = components.roleType;
        }
        return RelationshipMock;
    }());
    exports.RelationshipMock = RelationshipMock;
});
define("xrm-mock/url/reportopenparameters/reportopenparameters.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ReportOpenParametersMock = /** @class */ (function () {
        function ReportOpenParametersMock(action, id, helpID) {
            this.action = action;
            this.id = id;
            this.helpID = helpID;
        }
        return ReportOpenParametersMock;
    }());
    exports.ReportOpenParametersMock = ReportOpenParametersMock;
});
define("xrm-mock/retrievemultiplerequest/retrievemultiplerequest.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var RetrieveMultipleRequestMock = /** @class */ (function () {
        function RetrieveMultipleRequestMock(entities, nextLink) {
            this.entities = entities;
            this.nextLink = nextLink;
        }
        return RetrieveMultipleRequestMock;
    }());
    exports.RetrieveMultipleRequestMock = RetrieveMultipleRequestMock;
});
define("xrm-mock/events/saveeventarguments/saveeventarguments.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SaveEventArgumentsMock = /** @class */ (function () {
        function SaveEventArgumentsMock() {
        }
        SaveEventArgumentsMock.prototype.getSaveMode = function () {
            throw new Error("Not implemented.");
        };
        SaveEventArgumentsMock.prototype.isDefaultPrevented = function () {
            throw new Error("Not implemented.");
        };
        SaveEventArgumentsMock.prototype.preventDefault = function () {
            throw new Error("Not implemented.");
        };
        return SaveEventArgumentsMock;
    }());
    exports.SaveEventArgumentsMock = SaveEventArgumentsMock;
});
define("xrm-mock/events/saveeventcontext/saveeventcontext.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SaveEventContextMock = /** @class */ (function () {
        function SaveEventContextMock(eventContext) {
            this.eventContext = eventContext;
        }
        SaveEventContextMock.prototype.getContext = function () {
            return this.eventContext.getContext();
        };
        SaveEventContextMock.prototype.getDepth = function () {
            return this.eventContext.getDepth();
        };
        SaveEventContextMock.prototype.getEventSource = function () {
            return this.eventContext.getEventSource();
        };
        SaveEventContextMock.prototype.getFormContext = function () {
            return this.eventContext.getFormContext();
        };
        SaveEventContextMock.prototype.getSharedVariable = function (key) {
            return this.eventContext.getSharedVariable(key);
        };
        SaveEventContextMock.prototype.setSharedVariable = function (key, value) {
            return this.setSharedVariable(key, value);
        };
        SaveEventContextMock.prototype.getEventArgs = function () {
            throw new Error("Not implemented.");
        };
        return SaveEventContextMock;
    }());
    exports.SaveEventContextMock = SaveEventContextMock;
});
define("xrm-mock/saveoptions/saveoptions.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SaveOptionsMock = /** @class */ (function () {
        function SaveOptionsMock(UseSchedulingEngine) {
            this.UseSchedulingEngine = UseSchedulingEngine;
        }
        return SaveOptionsMock;
    }());
    exports.SaveOptionsMock = SaveOptionsMock;
});
define("xrm-mock/controls/section/section.mock", ["require", "exports", "xrm-mock/controls/uicangetvisibleelement/uicangetvisibleelement.mock", "xrm-mock/controls/uilabelelement/uilabelelement.mock", "xrm-mock/controls/uistandardelement/uistandardelement.mock"], function (require, exports, uicangetvisibleelement_mock_2, uilabelelement_mock_2, uistandardelement_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SectionMock = /** @class */ (function () {
        function SectionMock(name, parent, uiStandardElement, controls) {
            this.name = name;
            this.parent = parent;
            if (this.parent && this.parent.sections.get(name) == null) {
                var sections = this.parent.sections;
                if (sections) {
                    sections.push(this);
                }
            }
            this.uiStandardElement = uiStandardElement
                || new uistandardelement_mock_2.UiStandardElementMock(new uilabelelement_mock_2.UiLabelElementMock(name), new uicangetvisibleelement_mock_2.UiCanGetVisibleElementMock(true));
            this.controls = controls;
        }
        SectionMock.prototype.getName = function () {
            return this.name;
        };
        SectionMock.prototype.getParent = function () {
            return this.parent;
        };
        SectionMock.prototype.setVisible = function (visible) {
            this.uiStandardElement.setVisible(visible);
        };
        SectionMock.prototype.getVisible = function () {
            return this.uiStandardElement.getVisible();
        };
        SectionMock.prototype.getLabel = function () {
            return this.uiStandardElement.getLabel();
        };
        SectionMock.prototype.setLabel = function (label) {
            this.uiStandardElement.setLabel(label);
        };
        return SectionMock;
    }());
    exports.SectionMock = SectionMock;
});
define("xrm-mock/events/stagechangeeventarguments/stagechangeeventarguments.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StageChangeEventArgumentsMock = /** @class */ (function () {
        function StageChangeEventArgumentsMock() {
        }
        StageChangeEventArgumentsMock.prototype.getDirection = function () {
            throw new Error("Not implemented.");
        };
        StageChangeEventArgumentsMock.prototype.getStage = function () {
            throw new Error("Not implemented.");
        };
        return StageChangeEventArgumentsMock;
    }());
    exports.StageChangeEventArgumentsMock = StageChangeEventArgumentsMock;
});
define("xrm-mock/events/stagechangeeventcontext/stagechangeeventcontext.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StageChangeEventContextMock = /** @class */ (function () {
        function StageChangeEventContextMock(eventContext) {
            this.eventContext = eventContext;
        }
        StageChangeEventContextMock.prototype.getContext = function () {
            return this.eventContext.getContext();
        };
        StageChangeEventContextMock.prototype.getDepth = function () {
            return this.eventContext.getDepth();
        };
        StageChangeEventContextMock.prototype.getEventSource = function () {
            return this.eventContext.getEventSource();
        };
        StageChangeEventContextMock.prototype.getFormContext = function () {
            return this.eventContext.getFormContext();
        };
        StageChangeEventContextMock.prototype.getSharedVariable = function (key) {
            return this.eventContext.getSharedVariable(key);
        };
        StageChangeEventContextMock.prototype.setSharedVariable = function (key, value) {
            return this.setSharedVariable(key, value);
        };
        StageChangeEventContextMock.prototype.getEventArgs = function () {
            throw new Error("Not implemented.");
        };
        return StageChangeEventContextMock;
    }());
    exports.StageChangeEventContextMock = StageChangeEventContextMock;
});
define("xrm-mock/events/stageselectedeventarguments/stageselectedeventarguments.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StageSelectedEventArgumentsMock = /** @class */ (function () {
        function StageSelectedEventArgumentsMock() {
        }
        StageSelectedEventArgumentsMock.prototype.getStage = function () {
            throw new Error("Not implemented.");
        };
        return StageSelectedEventArgumentsMock;
    }());
    exports.StageSelectedEventArgumentsMock = StageSelectedEventArgumentsMock;
});
define("xrm-mock/events/stageselectedeventcontext/stageselectedeventcontext.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StageSelectedEventContextMock = /** @class */ (function () {
        function StageSelectedEventContextMock(eventContext) {
            this.eventContext = eventContext;
        }
        StageSelectedEventContextMock.prototype.getContext = function () {
            return this.eventContext.getContext();
        };
        StageSelectedEventContextMock.prototype.getDepth = function () {
            return this.eventContext.getDepth();
        };
        StageSelectedEventContextMock.prototype.getEventSource = function () {
            return this.eventContext.getEventSource();
        };
        StageSelectedEventContextMock.prototype.getFormContext = function () {
            return this.eventContext.getFormContext();
        };
        StageSelectedEventContextMock.prototype.getSharedVariable = function (key) {
            return this.eventContext.getSharedVariable(key);
        };
        StageSelectedEventContextMock.prototype.setSharedVariable = function (key, value) {
            return this.setSharedVariable(key, value);
        };
        StageSelectedEventContextMock.prototype.getEventArgs = function () {
            throw new Error("Not implemented.");
        };
        return StageSelectedEventContextMock;
    }());
    exports.StageSelectedEventContextMock = StageSelectedEventContextMock;
});
define("xrm-mock/processflow/step/step.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StepMock = /** @class */ (function () {
        function StepMock(name, attribute, required) {
            this.name = name;
            this.attribute = attribute;
            this.required = required;
        }
        StepMock.prototype.getAttribute = function () {
            return this.attribute;
        };
        StepMock.prototype.getName = function () {
            return this.name;
        };
        StepMock.prototype.isRequired = function () {
            return this.required;
        };
        return StepMock;
    }());
    exports.StepMock = StepMock;
});
define("xrm-mock/controls/stringcontrol/stringcontrol.mock", ["require", "exports", "xrm-mock/controls/autolookupcontrol/autolookupcontrol.mock"], function (require, exports, autolookupcontrol_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StringControlMock = /** @class */ (function (_super) {
        __extends(StringControlMock, _super);
        function StringControlMock(components) {
            return _super.call(this, components) || this;
        }
        return StringControlMock;
    }(autolookupcontrol_mock_2.AutoLookupControlMock));
    exports.StringControlMock = StringControlMock;
});
define("xrm-mock/attributes/stringattribute/stringattribute.mock", ["require", "exports", "xrm-mock/attributes/attribute/attribute.mock"], function (require, exports, attribute_mock_5) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var StringAttributeMock = /** @class */ (function (_super) {
        __extends(StringAttributeMock, _super);
        function StringAttributeMock(components) {
            var _this = _super.call(this, StringAttributeMock.defaultComponents(components)) || this;
            _this.format = components.format || "text";
            _this.maxLength = components.maxLength || 100;
            return _this;
        }
        StringAttributeMock.create = function (name, value) {
            return new StringAttributeMock({ name: name, value: value });
        };
        StringAttributeMock.defaultComponents = function (components) {
            components.attributeType = "string";
            return components;
        };
        StringAttributeMock.prototype.getFormat = function () {
            return _super.prototype.getFormat.call(this);
        };
        StringAttributeMock.prototype.getMaxLength = function () {
            return this.maxLength;
        };
        StringAttributeMock.prototype.setValue = function (value) {
            if (value && this.maxLength && value.length > this.maxLength) {
                throw new Error(("value cannot be greater than " + this.maxLength));
            }
            else {
                _super.prototype.setValue.call(this, value);
            }
        };
        return StringAttributeMock;
    }(attribute_mock_5.AttributeMock));
    exports.StringAttributeMock = StringAttributeMock;
});
define("xrm-mock/controls/tab/tab.mock", ["require", "exports", "xrm-mock-generator/helpers/array.helper", "xrm-mock/collection/itemcollection/itemcollection.mock", "xrm-mock/controls/uicangetvisibleelement/uicangetvisibleelement.mock", "xrm-mock/controls/uifocusable/uifocusable.mock", "xrm-mock/controls/uilabelelement/uilabelelement.mock", "xrm-mock/controls/uistandardelement/uistandardelement.mock"], function (require, exports, array_helper_2, itemcollection_mock_3, uicangetvisibleelement_mock_3, uifocusable_mock_2, uilabelelement_mock_3, uistandardelement_mock_3) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var TabMock = /** @class */ (function () {
        function TabMock(components) {
            var _this = this;
            this.uiStandardElement = components.uiStandardElement
                || new uistandardelement_mock_3.UiStandardElementMock(new uilabelelement_mock_3.UiLabelElementMock(components.name), new uicangetvisibleelement_mock_3.UiCanGetVisibleElementMock(true));
            this.uiFocusableElement = components.uiFocusableElement || new uifocusable_mock_2.UiFocusableMock(false);
            this.name = components.name;
            this.parent = components.parent;
            this.displayState = components.displayState || "expanded";
            this.tabStateChangeHandlers = components.tabStateChangeHandlers || [];
            this.sections = components.sections || new itemcollection_mock_3.ItemCollectionMock([]);
            this.sections.forEach(function (section, index) {
                var sectionMock = section;
                if (sectionMock) {
                    sectionMock.parent = _this;
                }
            });
        }
        TabMock.prototype.getDisplayState = function () {
            return this.displayState;
        };
        TabMock.prototype.getName = function () {
            return this.name;
        };
        TabMock.prototype.getParent = function () {
            return this.parent;
        };
        TabMock.prototype.setDisplayState = function (displayState) {
            this.displayState = displayState;
        };
        TabMock.prototype.setVisible = function (visible) {
            return this.uiStandardElement.setVisible(visible);
        };
        TabMock.prototype.getVisible = function () {
            return this.uiStandardElement.getVisible();
        };
        TabMock.prototype.getLabel = function () {
            return this.uiStandardElement.getLabel();
        };
        TabMock.prototype.setLabel = function (label) {
            return this.uiStandardElement.setLabel(label);
        };
        TabMock.prototype.setFocus = function () {
            return this.uiFocusableElement.setFocus();
        };
        TabMock.prototype.addTabStateChange = function (handler) {
            this.tabStateChangeHandlers.push(handler);
        };
        TabMock.prototype.removeTabStateChange = function (handler) {
            var index = array_helper_2.findIndex(this.tabStateChangeHandlers, handler);
            this.tabStateChangeHandlers.splice(index, 1);
        };
        return TabMock;
    }());
    exports.TabMock = TabMock;
});
define("xrm-mock/controls/timelinewall/timelinewall.mock", ["require", "exports", "xrm-mock/controls/control/control.mock"], function (require, exports, control_mock_3) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var TimelineWallMock = /** @class */ (function (_super) {
        __extends(TimelineWallMock, _super);
        function TimelineWallMock() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TimelineWallMock.prototype.refresh = function () {
            throw new Error("Not implemented.");
        };
        return TimelineWallMock;
    }(control_mock_3.ControlMock));
    exports.TimelineWallMock = TimelineWallMock;
});
define("xrm-mock/ui/ui.mock", ["require", "exports", "xrm-mock/collection/itemcollection/itemcollection.mock"], function (require, exports, itemcollection_mock_4) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var UiMock = /** @class */ (function () {
        function UiMock(components) {
            this.process = components.process;
            this.controls = components.controls || new itemcollection_mock_4.ItemCollectionMock([]);
            this.formSelector = components.formSelector;
            this.navigation = components.navigation;
            this.tabs = components.tabs || new itemcollection_mock_4.ItemCollectionMock([]);
            this.quickForms = components.quickForms || new itemcollection_mock_4.ItemCollectionMock([]);
        }
        UiMock.prototype.setFormNotification = function (message, level, uniqueId) {
            var formNotificationAlreadyExists = false;
            if (this.formNotifications && this.formNotifications.length) {
                formNotificationAlreadyExists = this._getFormNotificationExists(this.formNotifications, uniqueId);
            }
            if (formNotificationAlreadyExists) {
                return false;
            }
            else {
                if (this.formNotifications && this.formNotifications.length) {
                    this.formNotifications.push({ message: message, level: level, uniqueId: uniqueId });
                }
                else {
                    this.formNotifications = [{ message: message, level: level, uniqueId: uniqueId }];
                }
                return true;
            }
        };
        UiMock.prototype.clearFormNotification = function (uniqueId) {
            if (this.formNotifications && this.formNotifications.length) {
                var matchingNotificationsById = this.formNotifications.filter(function (item) {
                    return item.uniqueId === uniqueId;
                });
                if (matchingNotificationsById && matchingNotificationsById.length) {
                    var index = this.formNotifications.indexOf(matchingNotificationsById[0]);
                    this.formNotifications.splice(index, 1);
                    return true;
                }
                else {
                    return false;
                }
            }
            return false;
        };
        UiMock.prototype.close = function () {
            throw new Error(("close not implemented"));
        };
        UiMock.prototype.getFormType = function () {
            if (this.formSelector) {
                var currentForm = this.formSelector.getCurrentItem();
                if (currentForm) {
                    return currentForm.formType;
                }
            }
            else {
                return 0 /* Undefined */;
            }
        };
        UiMock.prototype.getViewPortHeight = function () {
            throw new Error(("getViewPortHeight not implemented"));
        };
        UiMock.prototype.getViewPortWidth = function () {
            throw new Error(("getViewPortWidth not implemented"));
        };
        UiMock.prototype.refreshRibbon = function () {
            throw new Error(("refreshRibbon not implemented"));
        };
        UiMock.prototype._getFormNotificationExists = function (notifications, uniqueId) {
            var matchingNotificationsById = notifications.filter(function (item) {
                return item.uniqueId === uniqueId;
            });
            return matchingNotificationsById && matchingNotificationsById.length ? true : false;
        };
        return UiMock;
    }());
    exports.UiMock = UiMock;
});
define("xrm-mock/usersettings/usersettings.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var UserSettingsMock = /** @class */ (function () {
        function UserSettingsMock(components) {
            this.defaultDashboardId = components.defaultDashboardId;
            this.isGuidedHelpEnabled = components.isGuidedHelpEnabled;
            this.isHighContrastEnabled = components.isHighContrastEnabled;
            this.isRTL = components.isRTL;
            this.languageId = components.languageId;
            this.securityRolePrivileges = components.securityRolePrivileges;
            this.securityRoles = components.securityRoles;
            this.transactionCurrencyId = components.transactionCurrencyId;
            this.userId = components.userId;
            this.userName = components.userName;
        }
        UserSettingsMock.prototype.dateFormattingInfo = function () {
            throw new Error("Not implemented.");
        };
        UserSettingsMock.prototype.getTimeZoneOffsetMinutes = function () {
            throw new Error("Not implemented");
        };
        return UserSettingsMock;
    }());
    exports.UserSettingsMock = UserSettingsMock;
});
define("xrm-mock/utility/utility.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var UtilityMock = /** @class */ (function () {
        function UtilityMock() {
        }
        UtilityMock.prototype.alertDialog = function (message, onCloseCallback) {
            alert(message);
            onCloseCallback();
        };
        UtilityMock.prototype.confirmDialog = function (message, yesCloseCallback, noCloseCallback) {
            if (confirm(message)) {
                yesCloseCallback();
            }
            else {
                noCloseCallback();
            }
        };
        UtilityMock.prototype.isActivityType = function (entityType) {
            throw new Error(("is activity type not implemented"));
        };
        UtilityMock.prototype.openQuickCreate = function (entityLogicalName, createFromEntity, parameters) {
            throw new Error(("open quick create not implemented"));
        };
        UtilityMock.prototype.openEntityForm = function (name, id, parameters, windowOptions) {
            throw new Error(("open entity form not implemented"));
        };
        UtilityMock.prototype.openWebResource = function (webResourceName, webResourceData, width, height) {
            throw new Error(("open web resource not implemented"));
        };
        UtilityMock.prototype.closeProgressIndicator = function () {
            throw new Error("Method not implemented.");
        };
        UtilityMock.prototype.getAllowedStatusTransitions = function (entityName, stateCode) {
            throw new Error("Method not implemented.");
        };
        UtilityMock.prototype.getEntityMetadata = function (entityName, attributes) {
            throw new Error("Method not implemented.");
        };
        UtilityMock.prototype.getGlobalContext = function () {
            throw new Error("Method not implemented.");
        };
        UtilityMock.prototype.getResourceString = function (webResourceName, key) {
            throw new Error("Method not implemented.");
        };
        UtilityMock.prototype.invokeProcessAction = function (name, parameters) {
            throw new Error("Method not implemented.");
        };
        UtilityMock.prototype.lookupObjects = function (lookupOptions) {
            throw new Error("Method not implemented.");
        };
        UtilityMock.prototype.refreshParentGrid = function (lookupOptions) {
            throw new Error("Method not implemented.");
        };
        UtilityMock.prototype.showProgressIndicator = function (message) {
            throw new Error("Method not implemented.");
        };
        return UtilityMock;
    }());
    exports.UtilityMock = UtilityMock;
});
define("xrm-mock/controls/viewselector/viewselector.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ViewSelectorMock = /** @class */ (function () {
        function ViewSelectorMock(isVisible) {
            this.visible = isVisible;
        }
        ViewSelectorMock.prototype.getCurrentView = function () {
            return this.currentView;
        };
        ViewSelectorMock.prototype.isVisible = function () {
            return this.visible;
        };
        ViewSelectorMock.prototype.setCurrentView = function (viewSelectorItem) {
            this.currentView = viewSelectorItem;
        };
        return ViewSelectorMock;
    }());
    exports.ViewSelectorMock = ViewSelectorMock;
});
define("xrm-mock/controls/viewselectoritem/viewselectoritem.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ViewSelectorItemMock = /** @class */ (function () {
        function ViewSelectorItemMock(reference) {
            this.reference = reference;
        }
        ViewSelectorItemMock.prototype.getEntityReference = function () {
            return this.reference;
        };
        return ViewSelectorItemMock;
    }());
    exports.ViewSelectorItemMock = ViewSelectorItemMock;
});
define("xrm-mock/webapi/webapi.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WebApiMock = /** @class */ (function () {
        function WebApiMock(clientContext, online, offline) {
            this.online = online;
            this.offline = offline;
            this.clientContext = clientContext;
            this.isOffline = clientContext.getClientState() === "Offline";
        }
        WebApiMock.prototype.createRecord = function (entityLogicalName, record) {
            if (this.isOffline) {
                return this.offline.createRecord(entityLogicalName, record);
            }
            else {
                return this.online.createRecord(entityLogicalName, record);
            }
        };
        WebApiMock.prototype.deleteRecord = function (entityLogicalName, id) {
            if (this.isOffline) {
                return this.offline.deleteRecord(entityLogicalName, id);
            }
            else {
                return this.online.deleteRecord(entityLogicalName, id);
            }
        };
        WebApiMock.prototype.retrieveRecord = function (entityLogicalName, id, options) {
            if (this.isOffline) {
                return this.offline.retrieveRecord(entityLogicalName, id, options);
            }
            else {
                return this.online.retrieveRecord(entityLogicalName, id, options);
            }
        };
        WebApiMock.prototype.retrieveMultipleRecords = function (entityLogicalName, options, maxPageSize) {
            if (this.isOffline) {
                return this.offline.retrieveMultipleRecords(entityLogicalName, options, maxPageSize);
            }
            else {
                return this.online.retrieveMultipleRecords(entityLogicalName, options, maxPageSize);
            }
        };
        WebApiMock.prototype.updateRecord = function (entityLogicalName, id, data) {
            if (this.isOffline) {
                return this.offline.updateRecord(entityLogicalName, id, data);
            }
            else {
                return this.online.updateRecord(entityLogicalName, id, data);
            }
        };
        WebApiMock.prototype.isAvailableOffline = function (entityLogicalName) {
            throw new Error("xrm-mock doesn\'t yet implement metadata features.");
        };
        return WebApiMock;
    }());
    exports.WebApiMock = WebApiMock;
});
define("xrm-mock/webapioffline/webapioffline.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WebApiOfflineMock = /** @class */ (function () {
        function WebApiOfflineMock() {
            this.notImplementedError = "WebApi methods not implemented. Consider faking API calls using a tool such as Sinon.JS";
        }
        WebApiOfflineMock.prototype.createRecord = function (entityLogicalName, record) {
            throw new Error(this.notImplementedError);
        };
        WebApiOfflineMock.prototype.deleteRecord = function (entityLogicalName, id) {
            throw new Error(this.notImplementedError);
        };
        WebApiOfflineMock.prototype.retrieveRecord = function (entityLogicalName, id, options) {
            throw new Error(this.notImplementedError);
        };
        WebApiOfflineMock.prototype.retrieveMultipleRecords = function (entityLogicalName, options, maxPageSize) {
            throw new Error(this.notImplementedError);
        };
        WebApiOfflineMock.prototype.updateRecord = function (entityLogicalName, id, data) {
            throw new Error(this.notImplementedError);
        };
        return WebApiOfflineMock;
    }());
    exports.WebApiOfflineMock = WebApiOfflineMock;
});
define("xrm-mock/webapionline/webapionline.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WebApiOnlineMock = /** @class */ (function () {
        function WebApiOnlineMock(offlineCapabilities) {
            this.notImplementedError = "WebApi methods not implemented. Consider faking API calls using a tool such as Sinon.JS";
            this.offline = offlineCapabilities;
        }
        WebApiOnlineMock.prototype.execute = function (request) {
            throw new Error(this.notImplementedError);
        };
        WebApiOnlineMock.prototype.executeMultiple = function (request) {
            throw new Error(this.notImplementedError);
        };
        WebApiOnlineMock.prototype.createRecord = function (entityLogicalName, record) {
            return this.offline.createRecord(entityLogicalName, record);
        };
        WebApiOnlineMock.prototype.deleteRecord = function (entityLogicalName, id) {
            return this.offline.deleteRecord(entityLogicalName, id);
        };
        WebApiOnlineMock.prototype.retrieveRecord = function (entityLogicalName, id, options) {
            return this.offline.retrieveRecord(entityLogicalName, id, options);
        };
        WebApiOnlineMock.prototype.retrieveMultipleRecords = function (entityLogicalName, options, maxPageSize) {
            return this.offline.retrieveMultipleRecords(entityLogicalName, options, maxPageSize);
        };
        WebApiOnlineMock.prototype.updateRecord = function (entityLogicalName, id, data) {
            return this.offline.updateRecord(entityLogicalName, id, data);
        };
        return WebApiOnlineMock;
    }());
    exports.WebApiOnlineMock = WebApiOnlineMock;
});
define("xrm-mock/xrmstatic.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var XrmStaticMock = /** @class */ (function () {
        function XrmStaticMock(components) {
            this.Device = components.device;
            this.Encoding = components.encoding;
            this.Mobile = components.mobile;
            this.Navigation = components.navigation;
            this.Page = components.page;
            this.Panel = components.panel;
            this.Utility = components.utility;
            this.WebApi = components.webApi;
        }
        return XrmStaticMock;
    }());
    exports.XrmStaticMock = XrmStaticMock;
});
define("xrm-mock/window.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WindowMock = /** @class */ (function () {
        function WindowMock(xrm) {
            this.Xrm = xrm;
        }
        WindowMock.prototype.GetGlobalContext = function () {
            throw new Error("not implemented");
        };
        return WindowMock;
    }());
    exports.WindowMock = WindowMock;
});
define("xrm-mock/index", ["require", "exports", "xrm-mock/controls/addcontrolnotificationoptions/addcontrolnotificationoptions.mock", "xrm-mock/navigation/alertstrings/alertstrings.mock", "xrm-mock/appproperties/appproperties.mock", "xrm-mock/metadata/attributemetadata/attributemetadata.mock", "xrm-mock/attributes/attribute/attribute.mock", "xrm-mock/controls/autocompletecommand/autocompletecommand.mock", "xrm-mock/controls/autocompleteresult/autocompleteresult.mock", "xrm-mock/controls/autocompleteresultset/autocompleteresultset.mock", "xrm-mock/controls/autolookupcontrol/autolookupcontrol.mock", "xrm-mock/attributes/booleanattribute/booleanattribute.mock", "xrm-mock/controls/booleancontrol/booleancontrol.mock", "xrm-mock/calendar/calendar.mock", "xrm-mock/device/capturefileresponse/capturefileresponse.mock", "xrm-mock/device/captureimageoptions/captureimageoptions.mock", "xrm-mock/clientcontext/clientcontext.mock", "xrm-mock/navigation/confirmresult/confirmresult.mock", "xrm-mock/navigation/confirmstrings/confirmstrings.mock", "xrm-mock/globalcontext/context.mock", "xrm-mock/controls/control/control.mock", "xrm-mock/controls/controlnotificationaction/controlnotificationaction.mock", "xrm-mock/data/data.mock", "xrm-mock/attributes/dateattribute/dateattribute.mock", "xrm-mock/controls/datecontrol/datecontrol.mock", "xrm-mock/dateformattinginfo/dateformattinginfo.mock", "xrm-mock/device/device.mock", "xrm-mock/navigation/dialogsizeoptions/dialogsizeoptions.mock", "xrm-mock/encoding/encoding.mock", "xrm-mock/navigation/entityformoptions/entityformoptions.mock", "xrm-mock/metadata/entitymetadata/entitymetadata.mock", "xrm-mock/entity/entity.mock", "xrm-mock/attributes/enumattribute/enumattribute.mock", "xrm-mock/navigation/errordialogoptions/errordialogoptions.mock", "xrm-mock/errorresponse/errorresponse.mock", "xrm-mock/events/eventcontext/eventcontext.mock", "xrm-mock/executeresponse/executeresponse.mock", "xrm-mock/navigation/filedetails/filedetails.mock", "xrm-mock/formcontext/formcontext.mock", "xrm-mock/controls/formitem/formitem.mock", "xrm-mock/utility/formopenparameters/formopenparameters.mock", "xrm-mock/controls/formselector/formselector.mock", "xrm-mock/device/getcurrentpositionresponse/getcurrentpositionresponse.mock", "xrm-mock/controls/gridcontrol/gridcontrol.mock", "xrm-mock/controls/grid/gridentity/gridentity.mock", "xrm-mock/controls/grid/grid.mock", "xrm-mock/controls/grid/gridrow/gridrow.mock", "xrm-mock/controls/grid/gridrowdata/gridrowdata.mock", "xrm-mock/collection/itemcollection/itemcollection.mock", "xrm-mock/metadata/label/label.mock", "xrm-mock/metadata/localizedlabel/localizedlabel.mock", "xrm-mock/attributes/lookupattribute/lookupattribute.mock", "xrm-mock/controls/lookupcontrol/lookupcontrol.mock", "xrm-mock/lookupoptions/lookupoptions.mock", "xrm-mock/lookupvalue/lookupvalue.mock", "xrm-mock/controls/navigation/navigation.mock", "xrm-mock/controls/navigationitem/navigationitem.mock", "xrm-mock/navigation/navigation.mock", "xrm-mock/attributes/numberattribute/numberattribute.mock", "xrm-mock/controls/numbercontrol/numbercontrol.mock", "xrm-mock/utility/openparameters/openparameters.mock", "xrm-mock/navigation/openwebresourceoptions/openwebresourceoptions.mock", "xrm-mock/attributes/optionsetattribute/optionsetattribute.mock", "xrm-mock/controls/optionsetcontrol/optionsetcontrol.mock", "xrm-mock/metadata/optionmetadata/optionmetadata.mock", "xrm-mock/optionsetvalue/optionsetvalue.mock", "xrm-mock/organizationsettings/organizationsettings.mock", "xrm-mock/page/page.mock", "xrm-mock/panel/panel.mock", "xrm-mock/device/pickfileoptions/pickfileoptions.mock", "xrm-mock/privilege/privilege.mock", "xrm-mock/controls/processcontrol/processcontrol.mock", "xrm-mock/processflow/processdictionary/processdictionary.mock", "xrm-mock/processflow/processmanager/processmanager.mock", "xrm-mock/processflow/process/process.mock", "xrm-mock/navigation/relationship/relationship.mock", "xrm-mock/url/reportopenparameters/reportopenparameters.mock", "xrm-mock/retrievemultiplerequest/retrievemultiplerequest.mock", "xrm-mock/events/saveeventarguments/saveeventarguments.mock", "xrm-mock/events/saveeventcontext/saveeventcontext.mock", "xrm-mock/saveoptions/saveoptions.mock", "xrm-mock/controls/section/section.mock", "xrm-mock/events/stagechangeeventarguments/stagechangeeventarguments.mock", "xrm-mock/events/stagechangeeventcontext/stagechangeeventcontext.mock", "xrm-mock/processflow/stage/stage.mock", "xrm-mock/events/stageselectedeventarguments/stageselectedeventarguments.mock", "xrm-mock/events/stageselectedeventcontext/stageselectedeventcontext.mock", "xrm-mock/controls/standardcontrol/standardcontrol.mock", "xrm-mock/processflow/step/step.mock", "xrm-mock/attributes/stringattribute/stringattribute.mock", "xrm-mock/controls/stringcontrol/stringcontrol.mock", "xrm-mock/controls/tab/tab.mock", "xrm-mock/controls/timelinewall/timelinewall.mock", "xrm-mock/controls/uicangetvisibleelement/uicangetvisibleelement.mock", "xrm-mock/controls/uifocusable/uifocusable.mock", "xrm-mock/controls/uikeypressable/uikeypressable.mock", "xrm-mock/controls/uilabelelement/uilabelelement.mock", "xrm-mock/ui/ui.mock", "xrm-mock/controls/uistandardelement/uistandardelement.mock", "xrm-mock/usersettings/usersettings.mock", "xrm-mock/utility/utility.mock", "xrm-mock/controls/viewselector/viewselector.mock", "xrm-mock/controls/viewselectoritem/viewselectoritem.mock", "xrm-mock/webapi/webapi.mock", "xrm-mock/webapioffline/webapioffline.mock", "xrm-mock/webapionline/webapionline.mock", "xrm-mock/window.mock", "xrm-mock/xrmstatic.mock"], function (require, exports, addcontrolnotificationoptions_mock_1, alertstrings_mock_1, appproperties_mock_1, attributemetadata_mock_1, attribute_mock_6, autocompletecommand_mock_1, autocompleteresult_mock_1, autocompleteresultset_mock_1, autolookupcontrol_mock_3, booleanattribute_mock_1, booleancontrol_mock_1, calendar_mock_1, capturefileresponse_mock_1, captureimageoptions_mock_1, clientcontext_mock_1, confirmresult_mock_1, confirmstrings_mock_1, context_mock_1, control_mock_4, controlnotificationaction_mock_1, data_mock_1, dateattribute_mock_1, datecontrol_mock_1, dateformattinginfo_mock_1, device_mock_1, dialogsizeoptions_mock_2, encoding_mock_1, entityformoptions_mock_1, entitymetadata_mock_1, entity_mock_1, enumattribute_mock_3, errordialogoptions_mock_1, errorresponse_mock_1, eventcontext_mock_1, executeresponse_mock_1, filedetails_mock_1, formcontext_mock_1, formitem_mock_1, formopenparameters_mock_1, formselector_mock_1, getcurrentpositionresponse_mock_1, gridcontrol_mock_1, gridentity_mock_1, grid_mock_1, gridrow_mock_1, gridrowdata_mock_1, itemcollection_mock_5, label_mock_1, localizedlabel_mock_1, lookupattribute_mock_1, lookupcontrol_mock_1, lookupoptions_mock_1, lookupvalue_mock_1, navigation_mock_1, navigationitem_mock_1, navigation_mock_2, numberattribute_mock_1, numbercontrol_mock_1, openparameters_mock_1, openwebresourceoptions_mock_1, optionsetattribute_mock_1, optionsetcontrol_mock_1, optionmetadata_mock_1, optionsetvalue_mock_2, organizationsettings_mock_1, page_mock_1, panel_mock_1, pickfileoptions_mock_1, privilege_mock_1, processcontrol_mock_1, processdictionary_mock_1, processmanager_mock_1, process_mock_1, relationship_mock_1, reportopenparameters_mock_1, retrievemultiplerequest_mock_1, saveeventarguments_mock_1, saveeventcontext_mock_1, saveoptions_mock_1, section_mock_1, stagechangeeventarguments_mock_1, stagechangeeventcontext_mock_1, stage_mock_1, stageselectedeventarguments_mock_1, stageselectedeventcontext_mock_1, standardcontrol_mock_6, step_mock_1, stringattribute_mock_1, stringcontrol_mock_1, tab_mock_1, timelinewall_mock_1, uicangetvisibleelement_mock_4, uifocusable_mock_3, uikeypressable_mock_2, uilabelelement_mock_4, ui_mock_1, uistandardelement_mock_4, usersettings_mock_1, utility_mock_1, viewselector_mock_1, viewselectoritem_mock_1, webapi_mock_1, webapioffline_mock_1, webapionline_mock_1, window_mock_1, xrmstatic_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AddControlNotificationOptionsMock = addcontrolnotificationoptions_mock_1.AddControlNotificationOptionsMock;
    exports.AlertStringsMock = alertstrings_mock_1.AlertStringsMock;
    exports.AppPropertiesMock = appproperties_mock_1.AppPropertiesMock;
    exports.AttributeMetadataMock = attributemetadata_mock_1.AttributeMetadataMock;
    exports.AttributeMock = attribute_mock_6.AttributeMock;
    exports.AutoCompleteCommandMock = autocompletecommand_mock_1.AutoCompleteCommandMock;
    exports.AutoCompleteResultMock = autocompleteresult_mock_1.AutoCompleteResultMock;
    exports.AutoCompleteResultSetMock = autocompleteresultset_mock_1.AutoCompleteResultSetMock;
    exports.AutoLookupControlMock = autolookupcontrol_mock_3.AutoLookupControlMock;
    exports.BooleanAttributeMock = booleanattribute_mock_1.BooleanAttributeMock;
    exports.BooleanControlMock = booleancontrol_mock_1.BooleanControlMock;
    exports.CalendarMock = calendar_mock_1.CalendarMock;
    exports.CaptureFileResponseMock = capturefileresponse_mock_1.CaptureFileResponseMock;
    exports.CaptureImageOptionsMock = captureimageoptions_mock_1.CaptureImageOptionsMock;
    exports.ClientContextMock = clientcontext_mock_1.ClientContextMock;
    exports.ConfirmResultMock = confirmresult_mock_1.ConfirmResultMock;
    exports.ConfirmStringsMock = confirmstrings_mock_1.ConfirmStringsMock;
    exports.ContextMock = context_mock_1.ContextMock;
    exports.ControlMock = control_mock_4.ControlMock;
    exports.ControlNotificationActionMock = controlnotificationaction_mock_1.ControlNotificationActionMock;
    exports.DataMock = data_mock_1.DataMock;
    exports.DateAttributeMock = dateattribute_mock_1.DateAttributeMock;
    exports.DateControlMock = datecontrol_mock_1.DateControlMock;
    exports.DateFormattingInfoMock = dateformattinginfo_mock_1.DateFormattingInfoMock;
    exports.DeviceMock = device_mock_1.DeviceMock;
    exports.DialogSizeOptionsMock = dialogsizeoptions_mock_2.DialogSizeOptionsMock;
    exports.EncodingMock = encoding_mock_1.EncodingMock;
    exports.EntityFormOptionsMock = entityformoptions_mock_1.EntityFormOptionsMock;
    exports.EntityMetadataMock = entitymetadata_mock_1.EntityMetadataMock;
    exports.EntityMock = entity_mock_1.EntityMock;
    exports.EnumAttributeMock = enumattribute_mock_3.EnumAttributeMock;
    exports.ErrorDialogOptionsMock = errordialogoptions_mock_1.ErrorDialogOptionsMock;
    exports.ErrorResponseMock = errorresponse_mock_1.ErrorResponseMock;
    exports.EventContextMock = eventcontext_mock_1.EventContextMock;
    exports.ExecuteResponseMock = executeresponse_mock_1.ExecuteResponseMock;
    exports.FileDetailsMock = filedetails_mock_1.FileDetailsMock;
    exports.FormContextMock = formcontext_mock_1.FormContextMock;
    exports.FormItemMock = formitem_mock_1.FormItemMock;
    exports.FormOpenParametersMock = formopenparameters_mock_1.FormOpenParametersMock;
    exports.FormSelectorMock = formselector_mock_1.FormSelectorMock;
    exports.GetCurrentPositionResponseMock = getcurrentpositionresponse_mock_1.GetCurrentPositionResponseMock;
    exports.GridControlMock = gridcontrol_mock_1.GridControlMock;
    exports.GridEntityMock = gridentity_mock_1.GridEntityMock;
    exports.GridMock = grid_mock_1.GridMock;
    exports.GridRowMock = gridrow_mock_1.GridRowMock;
    exports.GridRowDataMock = gridrowdata_mock_1.GridRowDataMock;
    exports.ItemCollectionMock = itemcollection_mock_5.ItemCollectionMock;
    exports.LabelMock = label_mock_1.LabelMock;
    exports.LocalizedLabelMock = localizedlabel_mock_1.LocalizedLabelMock;
    exports.LookupAttributeMock = lookupattribute_mock_1.LookupAttributeMock;
    exports.LookupControlMock = lookupcontrol_mock_1.LookupControlMock;
    exports.LookupOptionsMock = lookupoptions_mock_1.LookupOptionsMock;
    exports.LookupValueMock = lookupvalue_mock_1.LookupValueMock;
    exports.NavigationMock = navigation_mock_1.NavigationMock;
    exports.NavigationItemMock = navigationitem_mock_1.NavigationItemMock;
    exports.NavigationStaticMock = navigation_mock_2.NavigationStaticMock;
    exports.NumberAttributeMock = numberattribute_mock_1.NumberAttributeMock;
    exports.NumberControlMock = numbercontrol_mock_1.NumberControlMock;
    exports.OpenParametersMock = openparameters_mock_1.OpenParametersMock;
    exports.OpenWebresourceOptionsMock = openwebresourceoptions_mock_1.OpenWebresourceOptionsMock;
    exports.OptionSetAttributeMock = optionsetattribute_mock_1.OptionSetAttributeMock;
    exports.OptionSetControlMock = optionsetcontrol_mock_1.OptionSetControlMock;
    exports.OptionMetadataMock = optionmetadata_mock_1.OptionMetadataMock;
    exports.OptionSetValueMock = optionsetvalue_mock_2.OptionSetValueMock;
    exports.OrganizationSettingsMock = organizationsettings_mock_1.OrganizationSettingsMock;
    exports.PageMock = page_mock_1.PageMock;
    exports.PanelMock = panel_mock_1.PanelMock;
    exports.PickFileOptionsMock = pickfileoptions_mock_1.PickFileOptionsMock;
    exports.PrivilegeMock = privilege_mock_1.PrivilegeMock;
    exports.ProcessControlMock = processcontrol_mock_1.ProcessControlMock;
    exports.ProcessDictionaryMock = processdictionary_mock_1.ProcessDictionaryMock;
    exports.ProcessManagerMock = processmanager_mock_1.ProcessManagerMock;
    exports.ProcessMock = process_mock_1.ProcessMock;
    exports.RelationshipMock = relationship_mock_1.RelationshipMock;
    exports.ReportOpenParametersMock = reportopenparameters_mock_1.ReportOpenParametersMock;
    exports.RetrieveMultipleRequestMock = retrievemultiplerequest_mock_1.RetrieveMultipleRequestMock;
    exports.SaveEventArgumentsMock = saveeventarguments_mock_1.SaveEventArgumentsMock;
    exports.SaveEventContextMock = saveeventcontext_mock_1.SaveEventContextMock;
    exports.SaveOptionsMock = saveoptions_mock_1.SaveOptionsMock;
    exports.SectionMock = section_mock_1.SectionMock;
    exports.StageChangeEventArgumentsMock = stagechangeeventarguments_mock_1.StageChangeEventArgumentsMock;
    exports.StageChangeEventContextMock = stagechangeeventcontext_mock_1.StageChangeEventContextMock;
    exports.StageMock = stage_mock_1.StageMock;
    exports.StageSelectedEventArgumentsMock = stageselectedeventarguments_mock_1.StageSelectedEventArgumentsMock;
    exports.StageSelectedEventContextMock = stageselectedeventcontext_mock_1.StageSelectedEventContextMock;
    exports.StandardControlMock = standardcontrol_mock_6.StandardControlMock;
    exports.StepMock = step_mock_1.StepMock;
    exports.StringAttributeMock = stringattribute_mock_1.StringAttributeMock;
    exports.StringControlMock = stringcontrol_mock_1.StringControlMock;
    exports.TabMock = tab_mock_1.TabMock;
    exports.TimelineWallMock = timelinewall_mock_1.TimelineWallMock;
    exports.UiCanGetVisibleElementMock = uicangetvisibleelement_mock_4.UiCanGetVisibleElementMock;
    exports.UiFocusableMock = uifocusable_mock_3.UiFocusableMock;
    exports.UiKeyPressableMock = uikeypressable_mock_2.UiKeyPressableMock;
    exports.UiLabelElementMock = uilabelelement_mock_4.UiLabelElementMock;
    exports.UiMock = ui_mock_1.UiMock;
    exports.UiStandardElementMock = uistandardelement_mock_4.UiStandardElementMock;
    exports.UserSettingsMock = usersettings_mock_1.UserSettingsMock;
    exports.UtilityMock = utility_mock_1.UtilityMock;
    exports.ViewSelectorMock = viewselector_mock_1.ViewSelectorMock;
    exports.ViewSelectorItemMock = viewselectoritem_mock_1.ViewSelectorItemMock;
    exports.WebApiMock = webapi_mock_1.WebApiMock;
    exports.WebApiOfflineMock = webapioffline_mock_1.WebApiOfflineMock;
    exports.WebApiOnlineMock = webapionline_mock_1.WebApiOnlineMock;
    exports.WindowMock = window_mock_1.WindowMock;
    exports.XrmStaticMock = xrmstatic_mock_1.XrmStaticMock;
});
define("xrm-mock-generator/form", ["require", "exports", "xrm-mock/index"], function (require, exports, XrmMock) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Form = /** @class */ (function () {
        function Form() {
        }
        Form.createBlankForm = function () {
            var formItem = new XrmMock.FormItemMock({
                currentItem: true,
                formType: 1,
                id: "{00000000-0000-0000-0000-000000000000}",
                label: "",
            });
            return formItem;
        };
        return Form;
    }());
    exports.default = Form;
});
define("xrm-mock-generator/ui", ["require", "exports", "xrm-mock/index", "xrm-mock-generator/form"], function (require, exports, XrmMock, form_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Ui = /** @class */ (function () {
        function Ui() {
        }
        Ui.createUi = function () {
            var ui = new XrmMock.UiMock({
                controls: new XrmMock.ItemCollectionMock([]),
                formSelector: new XrmMock.FormSelectorMock(new XrmMock.ItemCollectionMock([form_1.default.createBlankForm()])),
                tabs: new XrmMock.ItemCollectionMock([]),
            });
            return ui;
        };
        Ui.createLabelElement = function (label) {
            return new XrmMock.UiLabelElementMock(label);
        };
        Ui.createCanGetVisibleElement = function (isVisible) {
            return new XrmMock.UiCanGetVisibleElementMock(isVisible);
        };
        Ui.createStandardElement = function (labelElement, visibleElement) {
            return new XrmMock.UiStandardElementMock(labelElement, visibleElement);
        };
        return Ui;
    }());
    exports.default = Ui;
});
define("xrm-mock-generator/control", ["require", "exports", "xrm-mock/index"], function (require, exports, XrmMock) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Control = /** @class */ (function () {
        function Control() {
        }
        Control.prototype.createBoolean = function (attributeOrComponents, name, visible, disabled, label) {
            if (visible === void 0) { visible = true; }
            if (disabled === void 0) { disabled = false; }
            var components = this.createStandardComponent(attributeOrComponents, name, visible, disabled, label);
            return this.addControl(new XrmMock.BooleanControlMock(components));
        };
        Control.prototype.createDate = function (attributeOrComponents, name, visible, disabled, label) {
            if (visible === void 0) { visible = true; }
            if (disabled === void 0) { disabled = false; }
            var components = this.createStandardComponent(attributeOrComponents, name, visible, disabled, label);
            return this.addControl(new XrmMock.DateControlMock(components));
        };
        Control.prototype.createGrid = function (nameOrComponents, visible, label) {
            if (visible === void 0) { visible = true; }
            var components = nameOrComponents.name
                ? nameOrComponents
                : {
                    label: label,
                    name: nameOrComponents,
                    visible: visible
                };
            return this.addControl(new XrmMock.GridControlMock(components));
        };
        Control.prototype.createLookup = function (attributeOrComponents, name, visible, disabled, label) {
            if (visible === void 0) { visible = true; }
            if (disabled === void 0) { disabled = false; }
            var components = this.createStandardComponent(attributeOrComponents, name, visible, disabled, label);
            return this.addControl(new XrmMock.LookupControlMock(components));
        };
        Control.prototype.createNumber = function (attributeOrComponents, name, visible, disabled, label) {
            if (visible === void 0) { visible = true; }
            if (disabled === void 0) { disabled = false; }
            var components = this.createStandardComponent(attributeOrComponents, name, visible, disabled, label);
            return this.addControl(new XrmMock.NumberControlMock(components));
        };
        Control.prototype.createOptionSet = function (attributeOrComponents, name, visible, disabled, label) {
            if (visible === void 0) { visible = true; }
            if (disabled === void 0) { disabled = false; }
            var components = this.createStandardComponent(attributeOrComponents, name, visible, disabled, label);
            return this.addControl(new XrmMock.OptionSetControlMock(components));
        };
        Control.prototype.createString = function (attributeOrComponents, name, visible, disabled, label) {
            var components = this.createStandardComponent(attributeOrComponents, name, visible, disabled, label);
            return this.addControl(new XrmMock.StringControlMock(components));
        };
        Control.prototype.addControl = function (control) {
            Xrm.Page.ui.controls.push(control);
            return control;
        };
        Control.prototype.createStandardComponent = function (attributeOrComponents, name, visible, disabled, label) {
            var att = attributeOrComponents;
            return att.controls && att.eventHandlers
                ? {
                    attribute: attributeOrComponents,
                    disabled: disabled,
                    label: label || name,
                    name: name,
                    visible: visible,
                }
                : attributeOrComponents;
        };
        return Control;
    }());
    exports.default = Control;
});
define("xrm-mock-generator/attribute", ["require", "exports", "xrm-mock/index", "xrm-mock-generator/control"], function (require, exports, XrmMock, control_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Attribute = /** @class */ (function () {
        function Attribute() {
            this.Control = new control_1.default();
        }
        Attribute.prototype.createBoolean = function (nameOrComponents, valueOrControlComponents) {
            if (typeof (nameOrComponents) === "string") {
                var components = { name: nameOrComponents, value: valueOrControlComponents };
                var controls = [{ name: nameOrComponents }];
                return this.associateAttribute(new XrmMock.BooleanAttributeMock(components), controls, "createBoolean");
            }
            else {
                return this.associateAttribute(new XrmMock.BooleanAttributeMock(nameOrComponents), this.arrayify(valueOrControlComponents), "createBoolean");
            }
        };
        Attribute.prototype.createDate = function (nameOrComponents, valueOrControlComponents) {
            if (typeof (nameOrComponents) === "string") {
                var components = { name: nameOrComponents, value: valueOrControlComponents };
                var controls = [{ name: nameOrComponents }];
                return this.associateAttribute(new XrmMock.DateAttributeMock(components), controls, "createDate");
            }
            else {
                return this.associateAttribute(new XrmMock.DateAttributeMock(nameOrComponents), this.arrayify(valueOrControlComponents), "createDate");
            }
        };
        Attribute.prototype.createLookup = function (nameOrComponents, valueOrControlComponents) {
            if (typeof (nameOrComponents) === "string") {
                var components = {
                    isPartyList: valueOrControlComponents && Array.isArray(valueOrControlComponents),
                    name: nameOrComponents,
                    value: this.arrayify(valueOrControlComponents)
                };
                var controls = [{ name: nameOrComponents }];
                return this.associateAttribute(new XrmMock.LookupAttributeMock(components), controls, "createLookup");
            }
            else {
                return this.associateAttribute(new XrmMock.LookupAttributeMock(nameOrComponents), this.arrayify(valueOrControlComponents), "createLookup");
            }
        };
        Attribute.prototype.createNumber = function (nameOrComponents, valueOrControlComponents) {
            if (typeof (nameOrComponents) === "string") {
                var components = { name: nameOrComponents, value: valueOrControlComponents };
                var controls = [{ name: nameOrComponents }];
                return this.associateAttribute(new XrmMock.NumberAttributeMock(components), controls, "createNumber");
            }
            else {
                return this.associateAttribute(new XrmMock.NumberAttributeMock(nameOrComponents), this.arrayify(valueOrControlComponents), "createNumber");
            }
        };
        Attribute.prototype.createOptionSet = function (nameOrComponents, valueOrControlComponents, options) {
            return typeof (nameOrComponents) === "string"
                ? this.createOptionSetFromParameters(nameOrComponents, valueOrControlComponents, options)
                : this.createOptionSetFromComponents(nameOrComponents, this.arrayify(valueOrControlComponents));
        };
        Attribute.prototype.createString = function (nameOrComponents, valueOrControlComponents) {
            if (valueOrControlComponents === void 0) { valueOrControlComponents = ""; }
            if (typeof (nameOrComponents) === "string") {
                var components = { name: nameOrComponents, value: valueOrControlComponents };
                var controls = [{ name: nameOrComponents }];
                return this.associateAttribute(new XrmMock.StringAttributeMock(components), controls, "createString");
            }
            else {
                return this.associateAttribute(new XrmMock.StringAttributeMock(nameOrComponents), this.arrayify(valueOrControlComponents), "createString");
            }
        };
        Attribute.prototype.createOptionSetFromParameters = function (name, value, options) {
            var num;
            if (value !== null
                && value !== undefined) {
                if (!options) {
                    options = [typeof value === "string"
                            ? { text: value, value: 0 }
                            : { text: value.toString(), value: value }];
                }
                if (typeof value === "string") {
                    var option = options.filter(function (o) { return o.text === value; })[0];
                    num = option.value;
                }
                else {
                    num = value;
                }
            }
            else {
                num = undefined;
            }
            var components = {
                name: name,
                options: options,
            };
            if (num || num === 0) {
                components.value = num;
            }
            var controls = [{ name: name, options: options }];
            return this.associateAttribute(new XrmMock.OptionSetAttributeMock(components), controls, "createOptionSet");
        };
        Attribute.prototype.createOptionSetFromComponents = function (components, controls) {
            if (components.options && components.options.length > 0) {
                controls.filter(function (c) { return !c.options; })
                    .forEach(function (c) {
                    c.options = components.options;
                });
            }
            return this.associateAttribute(new XrmMock.OptionSetAttributeMock(components), controls, "createOptionSet");
        };
        Attribute.prototype.createStringFromParameters = function (name, value) {
            var components = { name: name, value: value };
            var controls = [{ name: name }];
            return this.associateAttribute(new XrmMock.StringAttributeMock(components), controls, "createString");
        };
        Attribute.prototype.createAttribute = function (name, value) {
            var attribute = new XrmMock.AttributeMock({
                isDirty: false,
                name: name,
                submitMode: "dirty",
                value: value,
            });
            return attribute;
        };
        Attribute.prototype.addAttribute = function (attribute) {
            Xrm.Page.data.entity.attributes.push(attribute);
        };
        /**
         * Creates the given attribute, as well as the controls for the attribute defined by the components
         * @param attribute The newly created attribute to be added to the page colleciton of attributes
         * @param controls Array of Control Components to create controls for the given attribute
         * @param controlCreateFunction the name of the Control function to call to create the correct type of control
         */
        Attribute.prototype.associateAttribute = function (attribute, controls, controlCreateFunction) {
            var _this = this;
            this.addAttribute(attribute);
            controls.forEach(function (c) {
                c.attribute = attribute;
                _this.defaultName(c, attribute);
                _this.Control[controlCreateFunction](c);
            });
            return attribute;
        };
        Attribute.prototype.defaultName = function (control, attribute) {
            var names = [];
            attribute.controls.forEach(function (c) {
                names.push(c.getName());
            });
            if (!control.name) {
                control.name = attribute.getName();
            }
            else if (names.indexOf(control.name) >= 0) {
                throw new Error("Name " + control.name + " has already been defined for a control for attribute " + attribute.getName());
            }
            var i = 1;
            while (names.indexOf(control.name) >= 0) {
                control.name = attribute.getName() + i++;
            }
        };
        Attribute.prototype.arrayify = function (possibleArray) {
            if (!possibleArray) {
                return [];
            }
            else if (possibleArray instanceof Array) {
                return possibleArray;
            }
            else {
                return [possibleArray];
            }
        };
        return Attribute;
    }());
    exports.default = Attribute;
});
define("xrm-mock-generator/context", ["require", "exports", "xrm-mock/index"], function (require, exports, XrmMock) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Context = /** @class */ (function () {
        function Context() {
        }
        Context.createContext = function (client) {
            var context = new XrmMock.ContextMock({
                clientContext: new XrmMock.ClientContextMock(client || "Web", "Online"),
                userSettings: new XrmMock.UserSettingsMock({
                    isGuidedHelpEnabled: false,
                    isHighContrastEnabled: false,
                    isRTL: false,
                    userId: "1337",
                    userName: "jdoe",
                }),
            });
            return context;
        };
        return Context;
    }());
    exports.default = Context;
});
define("xrm-mock-generator/device", ["require", "exports", "xrm-mock/device/device.mock"], function (require, exports, device_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Device = /** @class */ (function () {
        function Device() {
        }
        Device.createDevice = function () {
            var device = new device_mock_2.DeviceMock();
            return device;
        };
        return Device;
    }());
    exports.default = Device;
});
define("xrm-mock-generator/formcontext", ["require", "exports", "xrm-mock/index", "xrm-mock-generator/ui"], function (require, exports, XrmMock, ui_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FormContext = /** @class */ (function () {
        function FormContext() {
        }
        FormContext.createFormContext = function (entity, ui, process) {
            return new XrmMock.FormContextMock(new XrmMock.DataMock(new XrmMock.EntityMock(entity), process), ui || ui_1.default.createUi());
        };
        return FormContext;
    }());
    exports.default = FormContext;
});
define("xrm-mock-generator/eventcontext", ["require", "exports", "xrm-mock/index", "xrm-mock-generator/context", "xrm-mock-generator/formcontext"], function (require, exports, XrmMock, context_1, formcontext_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var EventContext = /** @class */ (function () {
        function EventContext() {
        }
        EventContext.createEventContext = function (entity, context, formContext, ui, process) {
            var eventContext = new XrmMock.EventContextMock({
                context: context || context_1.default.createContext(),
                formContext: formContext || formcontext_1.default.createFormContext(entity, ui, process),
            });
            return eventContext;
        };
        EventContext.Context = new context_1.default();
        EventContext.FormContext = new formcontext_1.default();
        return EventContext;
    }());
    exports.default = EventContext;
});
define("xrm-mock/mobile/mobile.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var MobileMock = /** @class */ (function () {
        function MobileMock(offline) {
            this.offline = offline;
        }
        return MobileMock;
    }());
    exports.MobileMock = MobileMock;
});
define("xrm-mock/mobileoffline/mobileoffline.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var MobileOfflineMock = /** @class */ (function () {
        function MobileOfflineMock() {
        }
        MobileOfflineMock.prototype.isOfflineEnabled = function (entityType) {
            throw new Error(("not implemented"));
        };
        MobileOfflineMock.prototype.createRecord = function (entityType, data) {
            throw new Error(("not implemented"));
        };
        MobileOfflineMock.prototype.retrieveRecord = function (entityType, id, options) {
            throw new Error(("not implemented"));
        };
        MobileOfflineMock.prototype.retrieveMultipleRecords = function (entityType, options, maxPageSize) {
            throw new Error(("not implemented"));
        };
        MobileOfflineMock.prototype.updateRecord = function (entityType, id, data) {
            throw new Error(("not implemented"));
        };
        MobileOfflineMock.prototype.deleteRecord = function (entityType, id) {
            throw new Error(("not implemented"));
        };
        return MobileOfflineMock;
    }());
    exports.MobileOfflineMock = MobileOfflineMock;
});
define("xrm-mock-generator/mobile", ["require", "exports", "xrm-mock/mobile/mobile.mock", "xrm-mock/mobileoffline/mobileoffline.mock"], function (require, exports, mobile_mock_1, mobileoffline_mock_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Mobile = /** @class */ (function () {
        function Mobile() {
        }
        Mobile.createMobile = function () {
            var mobile = new mobile_mock_1.MobileMock(new mobileoffline_mock_1.MobileOfflineMock());
            return mobile;
        };
        return Mobile;
    }());
    exports.default = Mobile;
});
define("xrm-mock-generator/navigation", ["require", "exports", "xrm-mock/index"], function (require, exports, XrmMock) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Navigation = /** @class */ (function () {
        function Navigation() {
        }
        Navigation.createNavigation = function (client) {
            var navigation = new XrmMock.NavigationStaticMock();
            return navigation;
        };
        return Navigation;
    }());
    exports.default = Navigation;
});
define("xrm-mock-generator/helpers/control.helper", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ControlHelpers = /** @class */ (function () {
        function ControlHelpers() {
        }
        ControlHelpers.setControlsParent = function (controls, parent) {
            controls.forEach(function (control) {
                control.parent = parent;
            });
        };
        return ControlHelpers;
    }());
    exports.default = ControlHelpers;
});
define("xrm-mock-generator/section", ["require", "exports", "xrm-mock/controls/uicangetvisibleelement/uicangetvisibleelement.mock", "xrm-mock/controls/uilabelelement/uilabelelement.mock", "xrm-mock/controls/uistandardelement/uistandardelement.mock", "xrm-mock/index", "xrm-mock-generator/helpers/control.helper"], function (require, exports, uicangetvisibleelement_mock_5, uilabelelement_mock_5, uistandardelement_mock_5, XrmMock, control_helper_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Section = /** @class */ (function () {
        function Section() {
        }
        /**
         * Creates a section mock added to the XRM parent tab collection.
         * The section will be added to the parent tab given.
         * The section will be set as parent in the given controls.
         * Author: Yagasoft
         *
         * @param {string} [name] Logical name of the section.
         * @param {string} [label] Form label of the section.
         * @param {boolean} [isVisible] Is the section visible by default?
         * @param {Xrm.Controls.Tab} [parent] Parent tab.
         * @param {Xrm.Collection.ItemCollection<Xrm.Controls.Control>} [controls] List of controls in the section. Created by using the Mock Generator.
         * @returns {XrmMock.SectionMock} Section mock.
         * @memberof Section
         */
        Section.prototype.createSection = function (name, label, isVisible, parent, controls) {
            var section = new XrmMock.SectionMock(name, parent, new uistandardelement_mock_5.UiStandardElementMock(new uilabelelement_mock_5.UiLabelElementMock(label), new uicangetvisibleelement_mock_5.UiCanGetVisibleElementMock(isVisible)), controls);
            if (parent) {
                var sections = parent.sections;
                if (sections) {
                    sections.push(section);
                }
            }
            if (controls) {
                control_helper_1.default.setControlsParent(controls, section);
            }
            return section;
        };
        return Section;
    }());
    exports.default = Section;
});
define("xrm-mock-generator/tab", ["require", "exports", "xrm-mock/controls/uicangetvisibleelement/uicangetvisibleelement.mock", "xrm-mock/controls/uilabelelement/uilabelelement.mock", "xrm-mock/controls/uistandardelement/uistandardelement.mock", "xrm-mock/index", "xrm-mock-generator/helpers/control.helper"], function (require, exports, uicangetvisibleelement_mock_6, uilabelelement_mock_6, uistandardelement_mock_6, XrmMock, control_helper_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Tab = /** @class */ (function () {
        function Tab() {
        }
        /**
         * Creates a tab mock added to the XRM tabs collection.
         * The tab will be set as parent in the given sections.
         * Author: Yagasoft
         *
         * @param {string} [name] Logical name of the tab.
         * @param {string} [label] Form label of the tab.
         * @param {boolean} [isVisible] Is the tab visible by default?
         * @param {Xrm.DisplayState} [displayState] Is the tab collapsed or expanded?
         * @param {Xrm.ui} [parent] Parent UI.
         * @param {Xrm.Collection.ItemCollection<Xrm.Controls.Section>} [sections] List of sections in the tab. Created by using the Mock Generator.
         * @returns {XrmMock.TabMock} Tab mock.
         * @memberof Tab
         */
        Tab.prototype.createTab = function (name, label, isVisible, displayState, parent, sections) {
            var tabs = Xrm.Page.ui.tabs;
            if (!tabs) {
                throw new Error("Couldn't find the tabs collection in the XRM object.");
            }
            var tabComponents = {
                displayState: displayState,
                name: name,
                parent: parent || Xrm.Page.ui,
                sections: sections,
                uiStandardElement: new uistandardelement_mock_6.UiStandardElementMock(new uilabelelement_mock_6.UiLabelElementMock(label), new uicangetvisibleelement_mock_6.UiCanGetVisibleElementMock(isVisible)),
            };
            var tab = new XrmMock.TabMock(tabComponents);
            if (sections) {
                control_helper_2.default.setControlsParent(sections, tab);
            }
            tabs.push(tab);
            return tab;
        };
        return Tab;
    }());
    exports.default = Tab;
});
define("xrm-mock-generator/utility", ["require", "exports", "xrm-mock/utility/utility.mock"], function (require, exports, utility_mock_2) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Utility = /** @class */ (function () {
        function Utility() {
        }
        Utility.createUtility = function () {
            var utility = new utility_mock_2.UtilityMock();
            return utility;
        };
        return Utility;
    }());
    exports.default = Utility;
});
define("xrm-mock-generator/webapi", ["require", "exports", "xrm-mock/index"], function (require, exports, XrmMock) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WebApi = /** @class */ (function () {
        function WebApi() {
        }
        WebApi.createApi = function (clientContext) {
            var offlineApi = this.createOfflineApi();
            var onlineApi = this.createOnlineApi(offlineApi);
            var api = new XrmMock.WebApiMock(clientContext, onlineApi, offlineApi);
            return api;
        };
        WebApi.createOfflineApi = function () {
            var offlineApi = new XrmMock.WebApiOfflineMock();
            return offlineApi;
        };
        WebApi.createOnlineApi = function (offlineApi) {
            var onlineApi = new XrmMock.WebApiOnlineMock(offlineApi);
            return onlineApi;
        };
        return WebApi;
    }());
    exports.default = WebApi;
});
define("xrm-mock-generator/xrm-mock-generator", ["require", "exports", "xrm-mock/index", "xrm-mock-generator/attribute", "xrm-mock-generator/context", "xrm-mock-generator/control", "xrm-mock-generator/device", "xrm-mock-generator/eventcontext", "xrm-mock-generator/form", "xrm-mock-generator/formcontext", "xrm-mock-generator/mobile", "xrm-mock-generator/navigation", "xrm-mock-generator/section", "xrm-mock-generator/tab", "xrm-mock-generator/utility", "xrm-mock-generator/webapi"], function (require, exports, XrmMock, attribute_1, context_2, control_2, device_1, eventcontext_1, form_2, formcontext_2, mobile_1, navigation_1, section_1, tab_1, utility_1, webapi_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var XrmMockGenerator = /** @class */ (function () {
        function XrmMockGenerator() {
        }
        XrmMockGenerator.initialise = function (components) {
            components = components || {};
            this.context = components.context || context_2.default.createContext();
            this.formContext = formcontext_2.default.createFormContext(components.entity, components.ui, components.process);
            this.eventContext = eventcontext_1.default.createEventContext(components.entity, components.context, this.formContext, components.ui, components.process);
            var xrm = new XrmMock.XrmStaticMock({
                device: device_1.default.createDevice(),
                mobile: mobile_1.default.createMobile(),
                navigation: navigation_1.default.createNavigation(),
                page: new XrmMock.PageMock(this.context, this.formContext),
                utility: utility_1.default.createUtility(),
                webApi: webapi_1.default.createApi(this.context.client),
            });
            if (typeof global === "undefined") {
                window.Xrm = xrm;
            }
            else {
                global.Xrm = xrm;
            }
            return xrm;
        };
        XrmMockGenerator.getEventContext = function () {
            return this.eventContext;
        };
        XrmMockGenerator.getFormContext = function () {
            return this.formContext;
        };
        XrmMockGenerator.EventContext = new eventcontext_1.default();
        XrmMockGenerator.FormContext = new formcontext_2.default();
        XrmMockGenerator.Attribute = new attribute_1.default();
        XrmMockGenerator.Context = new context_2.default();
        XrmMockGenerator.Control = new control_2.default();
        XrmMockGenerator.Device = new device_1.default();
        XrmMockGenerator.Mobile = new mobile_1.default();
        XrmMockGenerator.Tab = new tab_1.default();
        XrmMockGenerator.Section = new section_1.default();
        XrmMockGenerator.Form = new form_2.default();
        XrmMockGenerator.Navigation = new navigation_1.default();
        XrmMockGenerator.WebApi = new webapi_1.default();
        XrmMockGenerator.Utility = new utility_1.default();
        return XrmMockGenerator;
    }());
    exports.XrmMockGenerator = XrmMockGenerator;
});
define("xrm-mock-generator/index", ["require", "exports", "xrm-mock-generator/xrm-mock-generator"], function (require, exports, xrm_mock_generator_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.XrmMockGenerator = xrm_mock_generator_1.XrmMockGenerator;
});
/// <amd-module name="xrm-mock"/>
define("xrm-mock", ["require", "exports", "xrm-mock/index", "xrm-mock-generator/index"], function (require, exports, index_1, index_2) {
    "use strict";
    function __export(m) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }
    Object.defineProperty(exports, "__esModule", { value: true });
    __export(index_1);
    exports.XrmMockGenerator = index_2.XrmMockGenerator;
});
define("xrm-mock/async/errorcallbackobject/errorcallbackobject.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ErrorCallbackObjectMock = /** @class */ (function () {
        function ErrorCallbackObjectMock(errorCode, message) {
            this.errorCode = errorCode;
            this.message = message;
        }
        return ErrorCallbackObjectMock;
    }());
    exports.ErrorCallbackObjectMock = ErrorCallbackObjectMock;
});
define("xrm-mock/async/offlineerrorcallbackobject/offlineerrorcallbackobject.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OfflineErrorCallbackObjectMock = /** @class */ (function () {
        function OfflineErrorCallbackObjectMock(errorCallbackObject, debugMessage) {
            this.errorCallbackObject = errorCallbackObject;
            this.debugMessage = debugMessage;
            this.errorCode = errorCallbackObject.errorCode;
            this.message = errorCallbackObject.message;
        }
        return OfflineErrorCallbackObjectMock;
    }());
    exports.OfflineErrorCallbackObjectMock = OfflineErrorCallbackObjectMock;
});
define("xrm-mock/async/offlineoperationsuccesscallbackobject/offlineoperationsuccesscallbackobject.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OfflineOperationSuccessCallbackObjectMock = /** @class */ (function () {
        function OfflineOperationSuccessCallbackObjectMock(id, logicalName) {
            this.id = id;
            this.logicalName = logicalName;
        }
        return OfflineOperationSuccessCallbackObjectMock;
    }());
    exports.OfflineOperationSuccessCallbackObjectMock = OfflineOperationSuccessCallbackObjectMock;
});
define("xrm-mock/async/openquickcreatesuccesscallbackobject/openquickcreatesuccesscallbackobject.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OpenQuickCreateSuccessCallbackObjectMock = /** @class */ (function () {
        function OpenQuickCreateSuccessCallbackObjectMock(savedEntityReference) {
            this.savedEntityReference = savedEntityReference;
        }
        return OpenQuickCreateSuccessCallbackObjectMock;
    }());
    exports.OpenQuickCreateSuccessCallbackObjectMock = OpenQuickCreateSuccessCallbackObjectMock;
});
define("xrm-mock/async/xrmpromise/xrmpromise.mock", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var XrmPromiseMock = /** @class */ (function () {
        function XrmPromiseMock() {
        }
        XrmPromiseMock.prototype.then = function (onFulfilled, onRejected) {
            throw new Error("Method not implemented.");
        };
        XrmPromiseMock.prototype.fail = function (onRejected) {
            throw new Error("Method not implemented.");
        };
        XrmPromiseMock.prototype.always = function (alwaysCallback) {
            throw new Error("Method not implemented.");
        };
        XrmPromiseMock.prototype.catch = function (onRejected) {
            throw new Error("Method not implemented.");
        };
        XrmPromiseMock.prototype.finally = function (finallyCallback) {
            throw new Error("Method not implemented.");
        };
        return XrmPromiseMock;
    }());
    exports.XrmPromiseMock = XrmPromiseMock;
});
